
# Application

## RcwlPipelines tools
### Rcwl scripts
The R scripts to build the CWL tools and pipelines based on the `Rcwl`
package are stored in the "src" folder with "tl_" and "pl_" prefix
respectively. The function `cwlTools` can be used to collect the
available scripts.  The `cachePath` can be your existing cache
directory or a new folder.

```{r}
tools <- cwlTools(cachePath = tempdir())
tools
```

The full paths can be pulled from the "fpath" column. The scripts can
be viewed to demonstrate how the tools and pipelines were built.
```{r}
library(dplyr)
library(conflicted)
## always use select from dplyr
conflict_prefer("select", "dplyr", quiet = TRUE)
conflict_prefer("filter", "dplyr", quiet = TRUE)
bfcinfo(tools) %>% select(rname, fpath)
```

The commands and docker containers from the wrapped tools are included
in the metadata.
```{r}
tls <- bfcinfo(tools) %>% filter(Type == "tool") %>%
    select(rname, Command, Container)
```

```{r, results = 'asis'}
knitr::kable(tls)
```

### Build a pipeline
We can develop a pipline by utilizing the available tools. For
example, a simple alignment pipelines with mapping and marking
duplicates can be built from the `tools`.

First, we check whether the required tools (bwa, samtools and picard
markduplicates) are available.

```{r}
bfcquery(tools, "bwa|sam2bam|sortBam|samtools_index|markdup") %>%
    filter(Type == "tool") %>%
    select(rname, Command, Container)
```

Next, we define the input parameters:
```{r}
p1 <- InputParam(id = "threads", type = "int")
p2 <- InputParam(id = "RG", type = "string")
p3 <- InputParam(id = "Ref", type = "string")
p4 <- InputParam(id = "FQ1", type = "File")
p5 <- InputParam(id = "FQ2", type = "File?")
```

Then we define the pipeline steps, from raw fastqs to duplicates
marked alignments:

```{r}
## bwa
s1 <- Step(id = "bwa", run = bwa,
           In = list(threads = "threads",
                     RG = "RG",
                     Ref = "Ref",
                     FQ1 = "FQ1",
                     FQ2 = "FQ2"))
## sam to bam
s2 <- Step(id = "sam2bam", run = sam2bam,
           In = list(sam = "bwa/sam"))
## sort bam
s3 <- Step(id = "sortBam", run = sortBam,
           In = list(bam = "sam2bam/bam"))
## mark duplicates
s4 <- Step(id = "markdup", run = markdup,
           In = list(ibam = "sortBam/sbam",
                     obam = list(
                         valueFrom="$(inputs.ibam.nameroot).mdup.bam"),
                     matrix = list(
                         valueFrom="$(inputs.ibam.nameroot).markdup.txt")))
## index bam
s5 <- Step(id = "idxBam", run = samtools_index,
           In = list(bam = "markdup/mBam"))
```

Last, we define the outputs and connect the steps to a new pipeline:
```{r}
req1 <- list(class = "StepInputExpressionRequirement")
req2 <- list(class = "InlineJavascriptRequirement")
## outputs
o1 <- OutputParam(id = "Bam", type = "File", outputSource = "markdup/mBam")
o2 <- OutputParam(id = "Idx", type = "File", outputSource = "idxBam/idx")
## stepParam
Align <- cwlStepParam(requirements = list(req1, req2),
                      inputs = InputParamList(p1, p2, p3, p4, p5),
                      outputs = OutputParamList(o1, o2))
## build pipeline
Align <- Align + s1 + s2 + s3 + s4 + s5
```

The pipeline is ready for use. We can plot the pipeline with
`plotCWL` from the `Rcwl` package.
```{r}
plotCWL(Align)
```

## RcwlPipelines summary
So far we have built 4 major pipelines in this package. Here is a
brief introduction to these 4 pipelines. More pipelines and tools are
expected to be included in the future.

## DNASeq alignment pipeline
The pipeline can be used to preprocess DNA sequences in fastq
format. It can take paired fastqs and read groups from multiple batches
as input.

```{r}
inputs(bwaMMRecal)
```

The pipeline includes two steps and several jobs will be run in each
step.

1. ``r names(runs(bwaMMRecal))[[1]]``: bwa alignment by read groups:
```{r}
runs(runs(bwaMMRecal)[[1]])
```
	
   * `bwa`: To align fastqs and read groups to reference genome with `bwa`.
   * `sam2bam`: To convert the alignments in "sam" format to "bam"
     format with `samtools`.
   * `sortBam`: To sort the "bam" file by coordinates with `samtools`.
   * `idxBam`: To index "bam" file with `samtools`.
  
2. ``r names(runs(bwaMMRecal))[[2]]``: To merge by samples and mark duplicates:
```{r}
runs(runs(bwaMMRecal)[[2]])
```

   * `mergeBam`: To merge bam files from multiple batches with `picard`.
   * `markdup`: To mark duplicates with `picard`.
   * `samtools_index`: To index bam file with `samtools`.
   * `samtools_flagstat`: To summarize flags in bam with `samtools`.

3. ``r names(runs(bwaMMRecal))[[3]]``: To apply TCGA
   BaseRecalibrator and ApplyBQSR.
```{r}
runs(runs(bwaMMRecal)[[3]])
```

   * `BaseRecalibrator`: To Generates recalibration table for Base
     Quality Score Recalibration with `gatk BaseRecalibrator`.
   * `ApplyBQSR`: To Apply base quality score recalibration with `gatk ApplyBQSR`.
   * `samtools_index`: To index bam file with `samtools`.
   * `samtools_flagstat`: To summarize flags in bam with `samtools`.
   * `samtools_stats`: To collects statistics from BAM file with
     `samtools`.

The final bam files with duplicates marked and base quality
recalibration, bam index, duplicates matrix, and statistics summaries
will be in the output folder.

```{r}
outputs(bwaMMRecal)
```

Here is the short summary and steps plot:
```{r}
short(bwaMMRecal)
plotCWL(bwaMMRecal)
```

### Prepare data
Here is a simple example of two samples. The "sample1" have two lanes of sequences and the "sample2" has only one pair of reads. The lists of reads1 `fq1`, reads2 `fq2`, read groups and output BAM names are defined in the `inputList`. The reference genome and number of threads to run the job are defined in the shared options, `paramList`:

```{r, eval=FALSE}
fq1 <- list(sample1 = list("apps/DNASeq/data/fq1_1.fq",
                           "apps/DNASeq/data/fq2_1.fq"),
            sample2 = list("apps/DNASeq/data/fq1_1.fq"))
fq2 <- list(sample1 = list("apps/DNASeq/data/fq1_2.fq",
                           "apps/DNASeq/data/fq2_2.fq"),
            sample2 = list("apps/DNASeq/data/fq1_2.fq"))
rgs <- list(sample1 = list("@RG\\tID:sample1.1\\tPL:Illumina\\tSM:sample1",
                           "@RG\\tID:sample1.2\\tPL:Illumina\\tSM:sample1"),
            sample2 = list("@RG\\tID:sample2.1\\tPL:Illumina\\tSM:sample2"))
samples <- list(sample1 = "sample1.bam", sample2 = "sample2.bam")

inputList <- list(outBam = samples, RG = rgs, FQ1s = fq1, FQ2s = fq2)
paramList <- list(threads = 2,
                  Ref = "apps/data/hs37d5.fa",
                  knowSites = list("apps/data/dbsnp_138.b37.vcf",
                                   "apps/data/Mills_and_1000G_gold_standard.indels.b37.vcf"))

```

### Run in cluster
```{r, eval=FALSE}
res <- runCWLBatch(bwaMMRecal, outdir = "apps/DNASeq/output/BAM",
                   inputList = inputList, paramList = paramList,
                   BPPARAM = BatchtoolsParam(workers = 2, cluster = "sge",
                                             template = "/rpcc/bioinformatics/sge_centos7.tmpl",
                                             resources = list(jobname="bwa",
                                                              threads = 2,
                                                              queue = "centos7.q"),
                                             log=TRUE, logdir=".", progressbar = T),
                   docker = FALSE, stderr = "")
```

List outputs:
```{r}
dir("apps/DNASeq/output/BAM/sample2")
```

## RNASeq pipeline
The pipeline was built with reads quality summary, `STAR` alignment,
quantification by `featureCounts` and `RSeQC` quality control. Here
are the inputs:

```{r}
inputs(rnaseq_Sf)
```

The pipeline includes 6 steps:

* `fastqc`: To run quality summary for raw fastqs with `fastqc`.
* `STAR`: To align fastqs with `STAR`.
* `samtools_index`: To index aligned bam file.
* `samtools_flagstat`: To summarize alignment flags.
* `featureCounts`: To quantify gene abundances.
* `RSeQC`: Several steps included.\
	- `gtfToGenePred`: To convert GTF annotation to "genePred" format.
	- `genePredToBed`: To convert "genePred" annotation to "bed" format.
	- `r_distribution`: To summarize reads distribution over genome features.
	- `gCoverage`: To summarize read coverage over gene body.

The outputs and logs from alignment, quantification and QC steps are
collected together into the output folder. A final QC report could be
generated by `multiqc`, which is also available in the data package.

Here are the short summary and steps plot:
```{r}
short(rnaseq_Sf)
plotCWL(rnaseq_Sf)
```

### Prepare data
An RNASeq test data set can be downloaded from [genomedata](http://genomedata.org/rnaseq-tutorial/HBR_UHR_ERCC_ds_5pc.tar), which includes paired-end fastqs for 6 samples.
```{r, eval=FALSE}
download.file("http://genomedata.org/rnaseq-tutorial/HBR_UHR_ERCC_ds_5pc.tar", "apps/RNASeq/data/HBR_UHR_ERCC_ds_5pc.tar)
untar("apps/RNASeq/data/HBR_UHR_ERCC_ds_5pc.tar", exdir = "apps/RNASeq/data/")
```

The input data must be in a named list, with the same names as the input list of the pipeline. For this pipeline, 5 inputs are required to be set, including `in_seqfiles`, `in_prefix`, `in_genomeDir`, `in_GTFfile` and `in_runThreadN`.

There are two different input lists, `inputList` and `paramList`. The `inputList` is used to define the inputs for each sample and will be submitted to different cluster nodes. The `paramList` is used to define parameters which are shared in all jobs.

Two following inputs should be listed in `inputList`:

* `in_seqfiles`: A list with the fastq files of each sample in each element. The names of the list are also required to be defined and can be the sample IDs. The length of the list will be the same as the number of samples, thus the list will be defined to `inputList` and assigned to different nodes for parallel computing.

* `in_prefix` is the same as `in_seqfiles`, which defines a list of sample IDs.

```{r}
files <- normalizePath(list.files("apps/RNASeq/data/", ".gz", full.names = TRUE))
files <- tapply(files, substring(basename(files), 1, 8), as.list)
inputList <- list(in_seqfiles = files,
                  in_prefix = as.list(names(files)))
```

These 3 parameter will be defined in `paramList`:

* `in_genomeDir`: The reference genome indexes for STAR.

* `in_GTFfile`: The gene annotation file in GTF format.

* `in_runThreadN`: The number of threads to run for each job.

```{r}
paramList <- list(
    in_genomeDir = "apps/data/GRCh38_100/",
    in_GTFfile = "apps/data/gencode.v25.annotation.gtf",
    in_runThreadN = 4
)
```

In some cases, we need to modify the default arguments in some steps of a pipeline. For example,
```{r}
arguments(rnaseq_Sf, "STAR")[[2]] <- "2"
head(arguments(rnaseq_Sf, "STAR"))
```

### Submit pipeline with SGE
The function `runCWLBatch` is used to submit the pipeline to cluster server. In addition to defining `inputList` and `paramList`, we need to define parallel parameters from the `BiocParallel` package. Here is an example where we use "sge" to submit the jobs. The "sge" template is a bash script with some predefined parameters for "qsub". The nodes queue name and number of slots/threads are variables from the template and can be assigned by the `resources` list. 

```{r, eval=FALSE}
res <- runCWLBatch(cwl = rnaseq_Sf, outdir = "apps/RNASeq/output/",
                   inputList = inputList, paramList = paramList,
                   BPPARAM = BatchtoolsParam(
                       workers = lengths(inputList)[1], cluster = "sge",
                       template = "apps/data/sge_centos7.tmpl",
                       resources = list(queue = "centos7.q",
                                        threads = 4)))
```

That's it! The fastqc files of each sample will be submitted to different nodes to run the whole pipeline automatically.

All the results have been collected to output directory of each sample. For example,
```{r}
dir("apps/RNASeq/output/HBR_Rep1")
```

### Summarize QC
The tool "multiqc" can aggregate results from the multiple outputs of the pipeline and generate a single page report, which also was implemented in the `RcwlPipelines` package:

```{r, eval=FALSE}
multiqc$dir <- "apps/RNASeq/output"
multiqc
```
We can also run the tool using `Rcwl` locally with the option `docker = TRUE`:

```{r, eval=FALSE}
runCWL(multiqc, stderr = "", Args = "--preserve-entire-environment", docker = FALSE)
```

**Here we got the QC report:**

<https://hubentu.github.io/others/multiqc_report.html>

## MC3 somatic variant calling pipeline
The Multi-Center Mutation Calling in Multiple Cancers project (MC3) pipeline was developed by TCGA to generate a comprehensive encyclopedia of somatic mutation calls. MC3 works by applying an ensemble of seven mutation-calling algorithms with scoring and artifact filtering. More details can be found in this paper:
[Scalable Open Science Approach for Mutation Calling of Tumor Exomes Using Multiple Genomic Pipelines](https://www.cell.com/fulltext/S2405-4712(18)30096-6)

The `mc3` pipeline is available at
<https://github.com/OpenGenomics/mc3>. All required software have been
deployed in cloud with docker.

The pipeline has been imported and contained in the `RcwlPipelines`
pacakge, which contains two major steps (markID step was removed):

- Call variants by 7 pipelines

- Merge VCF and convert to MAF

The steps of the pipeline was built on the CWL files from its github
repository, which were also contained in the package. Thereforce, we
need to load the pipleine by sourcing it from the script.

```{r}
bfcinfo(tools) %>% filter(rname == "mc3") %>% pull(rpath) %>% source
short(mc3)
plotCWL(mc3)
```

Two steps are included.\
1. `call_variants`: To call variants by 7 pipelines:
```{r}
callVar <- readCWL(runs(mc3)$call_variants)
plotCWL(callVar)
```

2. `covert`: To merge VCFs and convert to MAF:
```{r}
conv <- readCWL(runs(mc3)$convert)
plotCWL(conv)
```

The merged VCF and converted MAF files will be collected to the output
folder:

```{r}
outputs(mc3)
```

### Prepare data
Testing somatic mutation data can be download from: <https://github.com/genome/somatic-snv-test-data>.

1. Input list `inputList`.
The tumorID/normalID must be consistent with `SM` from BAM read group.
```{r}
inputList <- list(tumorID=list(test="NA12892"),
                  normalID=list(test="NA12878"),
                  tumor=list(test="apps/DNASeq/data/tumor.bam"),
                  normal=list(test="apps/DNASeq/data/normal.bam"))
```

2. Parameter list `paramList`.
```{r}
paramList <- list(bed_file="apps/data/mc3/gaf_20111020+broad_wex_1.1_hg19.bed",
                  centromere="apps/data/mc3/centromere_hg19.bed",
                  cosmic="apps/data/mc3/hg19_cosmic_v54_120711.vcf.gz",
                  dbsnp="apps/data/mc3/dbsnp_134_b37.leftAligned.vcf.gz",
                  refFasta="apps/data/human_g1k_v37.fa.gz",
                  vepData="apps/data/.vep/")
```

### Run MC3 pipeline
```{r, eval=FALSE}
res <- runCWLBatch(mc3, outdir = "apps/DNASeq/output/mc3",
                   inputList = inputList, paramList = paramList,
                   BPPARAM = BatchtoolsParam(workers = 1, cluster = "sge",
                                             template = "apps/data/sge_centos7.tmpl",
                                             resources = list(threads = 2,
                                                              queue = "centos7.q")))
```

The final VCF was filtered and merged from the outputs of different
callers and annotated by VEP:

```{r}
dir("apps/DNASeq/output/mc3/test")
```

The merged VCF file was converted to MAF file:
```{r}
vcf <- read.table("apps/DNASeq/output/mc3/test/merged.vep.vcf", sep="\t")
head(vcf)
```

## GATK4 germline variant calling pipeline
The GATK 4 best practice pipeline for germline variant calling was
implemented with Workflow Description Language (WDL), which is similar
to cwl and requires
[cromwell](https://github.com/broadinstitute/cromwell) to run the
pipeline. The details of the pipeline can be found here:
<https://software.broadinstitute.org/gatk/best-practices/workflow?id=11145> 

![Germline short variant discovery (SNPs + Indels)](https://us.v-cdn.net/5019796/uploads/editor/mz/tzm69d8e2spl.png)

The germline pipeline include 4 steps in WDL, paired fastq to ubam,
GATK alignment, variant calling by HaplotypeCaller and joint
genotyping. We wrapped the GATK pipeline into 3 steps using `Rcwl` for
different numbers of computing nodes requirements. The wrapped
pipeline can help to assign inputs to the input JSON templates and
glob results from the `cromwell` outputs.

1. `GAlign` GATK alignment.

The fastqs, sample information and customized json files for WDL are
required as inputs. Multiple steps will run in this step, including
`bwa` alignment, mark duplicates and base quality recalibration. GATK
ready BAM files will be collected into the output directory.

2. `hapCall` HaplotypeCaller.

The GATK ready BAM and customized json files are inputs in this
step. The local paths of GATK bundle files are required to be modified
in your json file. A "gVCF" files will be generated.

3. `jdCall` Joint variant discovery

This step will combine the "gVCF" files and then call germline
variants in all samples. The paths of the local bundle files are also
required to be changed in the json template file. The final VCF file
of germline variants will be generated.


### GATK Alignment
We wrapped the steps from raw fastqs to analysis-ready BAM file into `GAlign` pipeline. Here is the short summary of the pipeline.
```{r}
short(GAlign)
```

For the `inputList`, we need to assign the fastqs files and read groups for each sample. The inputs can be multiple items separated by comma if there are more than one read groups for each sample. The input templates and WDL scripts can be assigned in the `paramList`, and the reference and other GATK bundle files in the local json files should be changed accordingly to your local version of files. The path to the `cromwell` binary file is also required. Here is an example:

```{r, eval=FALSE}
tmpl1 <- system.file(package="RcwlPipelines", "GATK4/seq-format-conversion/paired-fastq-to-unmapped-bam.inputs.json")
tmpl2 <- system.file(package="RcwlPipelines", "GATK4/gatk4-data-processing/processing-for-variant-discovery-gatk4.hg38.wgs.inputs.local.json")
wdl1 <- system.file(package="RcwlPipelines", "GATK4/seq-format-conversion/paired-fastq-to-unmapped-bam.wdl")
wdl2 <- system.file(package="RcwlPipelines", "GATK4/gatk4-data-processing/processing-for-variant-discovery-gatk4.wdl")

inputList <- list(fastq1=list(normal="apps/DNASeq/data/normal_1.fq",
                              tumor="apps/DNASeq/data/tumor_1.fq"),
                  fastq2=list(normal="apps/DNASeq/data/normal_2.fq",
                              tumor="apps/DNASeq/data/tumor_2.fq"),
                  readGroup=list("normal.1", "tumor.1"),
                  sampleName=list("normal", "tumor"),
                  library=list("normal", "tumor"),
                  platunit=list("normal", "tumor"),
                  platform=list("illumina", "illumina"),
                  center=list("rpccc", "rpccc"))
paramList <- list(tmpl1=tmpl1,
                  wdl1=wdl1,
                  tmpl2=tmpl2,
                  wdl2=wdl2,
                  cromwell="/software/cromwell-36.jar")
r1 <- runCWLBatch(GAlign, outdir="apps/DNASeq/output/BAM", inputList, paramList,
                  BatchtoolsParam(workers = 2, cluster="sge",
                                  template = "apps/data/sge_centos7.tmpl",
                                  resources = list(threads = 16,
                                                   queue = "centos7.q")),
                  stderr="")
```

The outputs were globbed from the `cromwell` execution folder:
```{r}
list.files("apps/DNASeq/output/BAM/normal", recursive = TRUE)
```

### HaplotypeCaller
This step takes the BAM files as input and each BAM file will be assigned to different computing nodes. The json template file needs to be modified to include the correct GATK bundle paths first.
```{r, eval=FALSE}
wdl3 <- system.file(package="RcwlPipelines", "GATK4/gatk4-germline-snps-indels/haplotypecaller-gvcf-gatk4.wdl")
tmpl3 <- system.file(package="RcwlPipelines", "GATK4/gatk4-germline-snps-indels/haplotypecaller-gvcf-gatk4.hg38.inputs.local.json")

bams <- list(normal = normalizePath("output/BAM/normal/output/normal.hg38.bam"),
             tumor = normalizePath("output/BAM/tumor/output/tumor.hg38.bam"))
inputList <- list(bam = bams)
paramList <- list(intervals = normalizePath("output/interval.txt"),
                  cromwell = "/software/cromwell-36.jar",
                  wdl = wdl3,
                  tmpl = tmpl3)

r2 <- runCWLBatch(hapCall, outdir="apps/DNASeq/output/GATK", inputList, paramList,
                  BatchtoolsParam(workers = 2, cluster="sge",
                                  template = "apps/data/sge_centos7.tmpl",
                                  resources = list(threads = 16,
                                                   queue = "centos7.q")),
                  stderr="")

```

Here are the outputs:
```{r}
list.files("apps/DNASeq/output/GATK/normal", recursive = TRUE)
```


### Joint Discovery
The joint genotyping step will combine the `gvcf` files and then call variants in all samples, so only one computing node is required. Multiple values or files of the samples will need to be seperated by comma for each input in the `inputList`. The paths of the local bundle files will also need to be added to the json template file.
```{r, eval=FALSE}
wdl4 <- system.file(package="RcwlPipelines", "GATK4/gatk4-germline-snps-indels/joint-discovery-gatk4-local.wdl")
tmpl4 <- system.file(package="RcwlPipelines", "GATK4/gatk4-germline-snps-indels/joint-discovery-gatk4-local.hg38.wgs.inputs.json")

inputList <- list(sampleName = list(test="normal,tumor"),
                  gvcf = list(test="apps/DNASeq/output/GATK/normal/output/normal.hg38.g.vcf.gz,apps/DNASeq/output/GATK/tumor/output/tumor.hg38.g.vcf.gz"))

paramList <- list(callsetName = "test",
                  intervals = "apps/DNASeq/output/interval.21.interval_list",
                  unpadded_intervals = "apps/DNASeq/output/interval.21.intervals",
                  tmpl = tmpl4,
                  cromwell = "/software/cromwell-36.jar",
                  wdl = wdl4)

r3 <- runCWLBatch(jdCall, outdir="apps/DNASeq/output/GATK", inputList, paramList,
                  BatchtoolsParam(workers = 1, cluster="sge",
                                  template = "apps/sge_centos7.tmpl",
                                  resources = list(threads = 16,
                                                   queue = "centos7.q")),
                  stderr="")
```

Here are the final outputs:
```{r}
list.files("apps/DNASeq/output/GATK/test", recursive = TRUE)
```

