
# DNA-Seq Variant calling
First, we load the required packages.
```{r}
library(RcwlPipelines)
library(BiocParallel)
library(fs)
library(dplyr)
library(jsonlite)
library(conflicted)
## always use select from dplyr
conflict_prefer("select", "dplyr", quiet = TRUE)
```

A collection of resource bundle files, including reference, and variants files are required in most of the variant calling steps by different tools and pipelines. We collected a bundle of files in the data folder.
```{r}
ref <- "apps/data/hs37d5.fa"
dbsnp <- "apps/data/bundle/dbsnp_138.b37.vcf.gz"
```

## DNA alignment
We use the test data sets from the ICGC-TCGA DREAM Mutation Calling challenge for demonstration and performance test. The test sets contains two pairs of BAM files with SNVs and SVs in chr19 and chr20 separately. The 'Truth' VCF files are also included.

Here is the URL to download the test data.

* https://www.synapse.org/#!Synapse:syn2335184

```{r}
list.files("apps/data/DREAM/")
dir_info("apps/data/DREAM") %>% select(path, size)
```

In order to show the variant calling workflow from the very beginning. We convert the BAM files back to fastqs using `picard SamToFastq` and treat the two BAM files as samples from two patients.
```{r}
inputs(SamToFastq)
```

```{r, eval=FALSE}
## prepare inputs
bams <- list.files("apps/data/DREAM/", ".bam$", full.names = TRUE)
samples <- c("normal19", "tumor19", "normal20", "tumor20")
Bams <- tapply(bams, samples, as.list)
fq1 <- tapply(paste0(samples, "_R1.fq"), samples, as.list)
fq2 <- tapply(paste0(samples, "_R2.fq"), samples, as.list)

inputList <- list(bam = Bams,
                  fq1 = fq1,
                  fq2 = fq2)
## run in HPC using SGE
res1 <- runCWLBatch(SamToFastq, outdir = "apps/variants/fastq/", inputList,
                    BPPARAM = BatchtoolsParam(workers = length(samples), cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(jobname = "sam2fq",
                                                               queue = "all.q",
                                                               threads = 16),
                                              log=TRUE, logdir=".", progressbar = T),
                    stderr = "")
```

Let's check the outputs.
```{r}
dir_info("apps/variants/fastq/", recurse = TRUE, glob = "*.fq") %>% select(path, size)
```

Now, we are ready to run the DNA alignment pipeline, `bwaMRecal`, including all necessary steps to prepare the analysis-ready BAM files. We can check the required inputs and steps.
```{r}
inputs(bwaMRecal)
names(steps(bwaMRecal))
```

```{r, eval=FALSE}
fqs <- list.files("apps/variants/fastq", recursive = TRUE, full.names = TRUE)
fq1 <- fqs[grep("R1", fqs)]
fq2 <- fqs[grep("R2", fqs)]
ids <- sub("_.*", "", basename(fq1))
fq1L <- tapply(fq1, ids, as.list)
fq2L <- tapply(fq2, ids, as.list)

RGs <- paste("@RG",
             paste0("ID:", ids),
             paste0("LB:", ids),
             paste0("DT:", Sys.Date()),
             paste0("PL:", "Illumina"),
             "CN:RCWL",
             paste0("SM:", ids), sep = "\\t")
RGL <- tapply(RGs, ids, as.list)
outBam <- as.list(paste0(names(RGL), ".bam"))

inputList <- list(RG = RGL,
                  outBam = outBam,
                  FQ1s = fq1L,
                  FQ2s = fq2L)

paramList <- list(threads = 16,
                  Ref =ref,
                  knowSites = list(
                      "apps/data/bundle/dbsnp_138.b37.vcf",
                      "apps/data/bundle/Mills_and_1000G_gold_standard.indels.b37.vcf"
                  ))

res2 <- runCWLBatch(bwaMRecal, outdir = "apps/variants/BAM", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = length(samples),
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "align"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")

```

Here are the outputs:
```{r}
dir_info("apps/variants/BAM/", recurse = TRUE) %>% select(path, size)
```

## Germline variant calling
The GATK germline variant calling pipeline is one of the most popular method to call germline short variants. The latest best practice was implemented with the Workflow Description Language (WDL). The WDL scripts was simply wrapped with `Rcwl` in the `RcwlPipelines` package.

* https://github.com/gatk-workflows/gatk4-germline-snps-indels

There are two steps to run the latest GATK pipeline if the BAMs are prepared. First the `haplotypecaller` step is used to call variants by samples. The "WDL" script and "json" input for this step are also contained in the package.
```{r}
wdl1 <- system.file(package="RcwlPipelines", "GATK4/gatk4-germline-snps-indels/haplotypecaller-gvcf-gatk4.wdl")
tmpl1 <- system.file(package="RcwlPipelines", "GATK4/gatk4-germline-snps-indels/haplotypecaller-gvcf-gatk4.hg38.inputs.local.json")
```

We need to modify the reference from "hg38" to "b37". Please note that all the file paths should be absolute paths for the "WDL" inputs.
```{r, eval=FALSE}
json1 <- fromJSON(tmpl1)
json1$HaplotypeCallerGvcf_GATK4.ref_dict <- normalizePath("apps/data/hs37d5.dict")
json1$HaplotypeCallerGvcf_GATK4.ref_fasta <- normalizePath("apps/data/hs37d5.fa")
json1$HaplotypeCallerGvcf_GATK4.ref_fasta_index <- normalizePath("apps/data/hs37d5.fa.fai")
dir.create("apps/variants/vcf")
writeLines(jsonlite::toJSON(json1, pretty = TRUE, auto_unbox = T), "apps/variants/vcf/tmpl1.json")
```

A interval list is required to define the exome capture region. We couldn't find one for the TCGA data, so here we just simply use the exome regions.
```{r, eval=FALSE}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
options(scipen = 999)
e1 <- reduce(exons(TxDb.Hsapiens.UCSC.hg19.knownGene), ignore.strand = TRUE)
e1 <- e1[seqnames(e1) %in% c("chr19", "chr20")]
e1df <- data.frame(e1)
bed <- data.frame(as.character(sub("chr", "", e1df[,1])), e1df[,2] - 1, e1df[,3])
write.table(bed, "apps/variants/vcf/intval.bed", sep = "\t",
            row.names = FALSE, col.names = FALSE, quote = FALSE)

BedToIntervalList$bed <- "apps/variants/vcf/intval.bed"
BedToIntervalList$SD <- "apps/data/hs37d5.dict"
BedToIntervalList$out <- "region.interval_list"
r1 <- runCWL(BedToIntervalList, outdir = "apps/variants/vcf/")

writeLines(normalizePath("apps/variants/vcf/region.interval_list"), "apps/variants/vcf/intval.txt")
fai <- read.table("apps/data/hs37d5.fa.fai")
unpad <- paste0(c(19, 20), ":", 1, "-", fai[19:20, 2])
writeLines(unpad, "apps/variants/vcf/intval.unpad.txt")
```

```{r, eval=FALSE}
## fix bai paths
file.copy("apps/variants/BAM/normal19/normal19.bam.bai",
          "apps/variants/BAM/normal19/normal19.bai")
file.copy("apps/variants/BAM/normal20/normal20.bam.bai",
          "apps/variants/BAM/normal20/normal20.bai")

bams <- normalizePath(list.files("apps/variants/BAM", "normal.*.bam$",
                                 recursive = TRUE, full.names = TRUE))
bams <- as.list(bams)
names(bams) <- c("normal19", "normal20")

inputList <- list(bam = bams)
paramList <- list(intervals = normalizePath("apps/variants/vcf/intval.txt"),
                  cromwell = normalizePath("apps/cromwell-45.jar"),
                  wdl = wdl1,
                  tmpl = normalizePath("apps/variants/vcf/tmpl1.json"))

res3 <- runCWLBatch(hapCall, outdir = "apps/variants/vcf", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = length(bams),
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "hapCall"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "", cwlTemp=T)
```

Here are the results.
```{r}
dir_info("apps/variants/vcf/", recurse = TRUE, glob = "*.g.vcf.gz") %>% select(path, size)
```

Then we perform joint genotyping with `jdCall`. The reference genome bundles are also need to be changed to "b37". You can skip this step if you already have a "b37" version of metadata file.
```{r, eval=FALSE}
wdl2 <- system.file(package="RcwlPipelines", "GATK4/gatk4-germline-snps-indels/joint-discovery-gatk4-local.wdl")
tmpl2 <- system.file(package="RcwlPipelines", "GATK4/gatk4-germline-snps-indels/joint-discovery-gatk4-local.hg38.wgs.inputs.json")
## change to b37
json2 <- fromJSON(tmpl2)
json2$JointGenotyping.ref_dict <- normalizePath("apps/data/hs37d5.dict")
json2$JointGenotyping.ref_fasta <- normalizePath("apps/data/hs37d5.fa")
json2$JointGenotyping.ref_fasta_index <- normalizePath("apps/data/hs37d5.fa.fai")
json2$JointGenotyping.dbsnp_vcf <- normalizePath("apps/data/bundle/dbsnp_138.b37.vcf")
json2$JointGenotyping.dbsnp_vcf_index <- normalizePath("apps/data/bundle/dbsnp_138.b37.vcf.idx")
json2$JointGenotyping.one_thousand_genomes_resource_vcf <- normalizePath("apps/data/bundle/1000G_phase1.snps.high_confidence.b37.vcf")
json2$JointGenotyping.one_thousand_genomes_resource_vcf_index <- normalizePath("apps/data/bundle/1000G_phase1.snps.high_confidence.b37.vcf.idx")
json2$JointGenotyping.omni_resource_vcf <- normalizePath("apps/data/bundle/1000G_omni2.5.b37.vcf")
json2$JointGenotyping.omni_resource_vcf_index <- normalizePath("apps/data/bundle/1000G_omni2.5.b37.vcf.idx")
json2$JointGenotyping.mills_resource_vcf <- normalizePath("apps/data/bundle/Mills_and_1000G_gold_standard.indels.b37.vcf")
json2$JointGenotyping.mills_resource_vcf_index <- normalizePath("apps/data/bundle/Mills_and_1000G_gold_standard.indels.b37.vcf.idx")
json2$JointGenotyping.axiomPoly_resource_vcf <- normalizePath("apps/data/bundle/Axiom_Exome_Plus.genotypes.all_populations.poly.vcf.gz")
json2$JointGenotyping.axiomPoly_resource_vcf_index <- normalizePath("apps/data/bundle/Axiom_Exome_Plus.genotypes.all_populations.poly.vcf.gz.tbi")
json2$JointGenotyping.hapmap_resource_vcf <- normalizePath("apps/data/bundle/hapmap_3.3.b37.vcf")
json2$JointGenotyping.hapmap_resource_vcf_index <- normalizePath("apps/data/bundle/hapmap_3.3.b37.vcf.idx")

writeLines(toJSON(json2, pretty = TRUE, auto_unbox = T), "apps/variants/vcf/tmpl2.json")
```

The joint calling step takes the inputs from the previous step and run together.
```{r, eval=FALSE}
## prepare inputs
sampleName <- paste(names(bams), collapse = ",")
gvcf <- paste(normalizePath(list.files("apps/variants/vcf", "g.vcf.gz$",
                                       recursive = TRUE, full.names = TRUE)),
              collapse = ",")

inputList <- list(sampleName = list(res = sampleName),
                  gvcf = list(res = gvcf))

paramList <- list(callsetName = "germline",
                  intervals = normalizePath("apps/variants/vcf/region.interval_list"),
                  unpadded_intervals = normalizePath("apps/variants/vcf/intval.unpad.txt"),
                  tmpl = normalizePath("apps/variants/vcf/tmpl2.json"),
                  wdl = wdl2,
                  cromwell = normalizePath("apps/cromwell-45.jar"))
res4 <- runCWLBatch(jdCall, outdir = "apps/variants/vcf1", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 1,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "jdCall"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "", cwlTemp=T)
```

Here are the final results:
```{r}
dir_info("apps/variants/vcf/res/output", recurse = TRUE) %>% select(path, size)
```

## Somatic mutation calling
We have collected many somatic mutation callers, including MuTect2, strelka2, MuSe, VarDict, SomaticSniper, LoFreq, VarScan2, lancet and neusomatic. Recent research papers have reported that the combination of multiple callers could improve the sensitivity and specificity.

### MuTect2
The MuTect2 was built based on the latest best practise from GATK 4.
<https://software.broadinstitute.org/gatk/best-practices/workflow?id=11146>

The required resources files can be downloaded from its google storage. <gs://gatk-best-practices>

1. Call variants on normal samples

```{r, eval=FALSE}
bams <- list.files("apps/variants/BAM", ".bam$", recursive = TRUE, full.names = TRUE)
nbam <- as.list(bams[1:2])
names(nbam) <- c("normal19", "normal20")
ovcf <- as.list(paste0(names(nbam), ".vcf"))

inputList <- list(tbam = nbam,
                  out = ovcf)
paramList <- list(Ref = ref,
                  interval = "apps/variants/vcf/intval.bed")
## avoid a bug in GenomicsDBImport
arguments(Mutect2) <- list("-max-mnp-distance", "0")
res5 <- runCWLBatch(Mutect2, outdir = "apps/variants/mutect2", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "mutect2"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")

```

```{r}
dir_info("apps/variants/mutect2", regexp = "normal", recurse = TRUE) %>% select(path, size)
```

Note that, we need to add "-max-mnp-distance 0" to the arguments according to the Mutect2 documents to avoid a bug in "GenomicsDBImport" step.


2. Create a panel of normals (PoN)
```{r, eval=FALSE}
GPoN$nvcf <- as.list(list.files("apps/variants/mutect2/", ".vcf$",
                                recursive = TRUE, full.names = TRUE))
GPoN$Ref <- ref
GPoN$interval <- "apps/variants/vcf/intval.bed"
GPoN$pvcf <- "pon.vcf"
##GPoN$gresource <- "apps/data/Mutect2/af-only-gnomad.raw.sites.b37.vcf"
runCWL(GPoN, outdir = "apps/variants/mutect2/", stderr = "", Args = "--relax-path-checks")
```

Here we add "--relax-path-checks" arugment to the "cwltool" because the temporary files have special characters in their file names. 

3. Mutect2 pipeline
We can visulize how the pipeline `Mutect2PL` was built based on the best practice guide.
```{r}
plotCWL(Mutect2PL)
```

```{r, eval=FALSE}
tbam <- as.list(bams[3:4])
nbam <- as.list(bams[1:2])
nid <- as.list(sub(".bam", "", basename(bams))[1:2])
tid <- as.list(sub(".bam", "", basename(bams))[3:4])

names(tbam) <- names(nbam) <- c("pt19", "pt20")
inputList <- list(tbam = tbam, nbam = nbam,
                  normal = nid, tumor = tid)
paramList <- list(Ref = ref,
                  gresource = "apps/data/Mutect2/af-only-gnomad.raw.sites.b37.vcf",
                  pon = "apps/variants/mutect2/pon.vcf",
                  interval = "apps/variants/vcf/intval.bed",
                  comvcf = "apps/data/Mutect2/GetPileupSummaries/small_exac_common_3_b37.vcf")

res6 <- runCWLBatch(Mutect2PL, outdir = "apps/variants/mutect2", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "mutect2"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")

```

Here are the final results for sample "pt19":
```{r}
list.files("apps/variants/mutect2/pt19")
```

### MuSE
Only one step is required to run `MuSE`. We only need to prepare the BAM files, reference genome, caputre region and the dbSNP vcf file.
```{r, eval=FALSE}
inputList <- list(tbam = tbam,
                  nbam = nbam,
                  vcf = as.list(paste0(names(tbam), "_MuSE.vcf")))
paramList <- list(ref = ref,
                  region = "apps/variants/vcf/intval.bed",
                  dbsnp = dbsnp)

res7 <- runCWLBatch(MuSE, outdir = "apps/variants/MuSE", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "MuSE"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")
```

Here are the results:
```{r}
dir_info("apps/variants/MuSE", glob = "*.vcf", recurse = TRUE) %>% select(path, size)
```

More details about the tool can be found:
<https://bioinformatics.mdanderson.org/public-software/muse/>.

### Strelka2
The tool recommended to use the indel candidates from `manta` as input for `strelka2`. We combined the two tools as `mantaStrelka` pipeline. 

The tool also recommended to use a indexed region file.
```{r, eval=FALSE}
bgzip$ifile <- "apps/variants/vcf/intval.bed"
runCWL(bgzip, outdir = "apps/variants/vcf/", stderr = "")
tabix_index$tfile <- "apps/variants/vcf/intval.bed.gz"
tabix_index$type <- "bed"
runCWL(tabix_index, outdir = "apps/variants/vcf/", stderr = "")
```

To prepare the inputs and run it.
```{r, eval=FALSE}
inputList <- list(tbam = tbam,
                  nbam = nbam)
paramList <- list(ref = ref,
                  region = "apps/variants/vcf/intval.bed.gz")

res8 <- runCWLBatch(mantaStrelka, outdir = "apps/variants/strelka", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "strelka"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")
```

Here are the results:
```{r}
dir_info("apps/variants/strelka", glob = "*.vcf.gz*", recurse = TRUE) %>% select(path, size)
```

More details about the tool can be found:
<https://github.com/Illumina/strelka>

### SomaticSniper
The `SomaticSniper` only requires bams and reference genome to run.
```{r, eval=FALSE}
inputList <- list(tbam = tbam,
                  nbam = nbam,
                  vcf = as.list(paste0(names(tbam), "_SomaticSniper.vcf")))
paramList <- list(ref = ref)

res9 <- runCWLBatch(SomaticSniper, outdir = "apps/variants/SomaticSniper", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "SomaticSniper"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")
```

Here are the results:
```{r}
dir_info("apps/variants/SomaticSniper", glob = "*.vcf", recurse = TRUE) %>% select(path, size)
```

More details can be found:
<https://github.com/genome/somatic-sniper>

### VarDict
Similar to the other tools, we only need to prepare BAM files, reference genome and the capture region file.
```{r, eval=FALSE}
inputList <- list(tbam = tbam,
                  nbam = nbam,
                  vcf = as.list(paste0(names(tbam), "_VarDict.vcf")))
paramList <- list(ref = ref,
                  region = "apps/variants/vcf/intval.bed")

res10 <- runCWLBatch(VarDict, outdir = "apps/variants/VarDict", inputList, paramList,
                     BPPARAM = BatchtoolsParam(workers = 2,
                                               cluster = "sge",
                                               template = "apps/data/sge.tmpl",
                                               resources = list(threads = 16,
                                                                queue = "all.q",
                                                                jobname = "VarDict"),
                                               log = TRUE, logdir = ".", progressbar = TRUE),
                     stderr = "")
```

Here are the results:
```{r}
dir_info("apps/variants/VarDict", glob = "*.vcf", recurse = TRUE) %>% select(path, size)
```

More details can be found:
<https://github.com/AstraZeneca-NGS/VarDict>

### LoFreq
The tools requires BAM files, reference genome, dbSNP and capture regions.
```{r, eval=FALSE}
inputList <- list(tbam = tbam,
                  nbam = nbam,
                  out = as.list(paste0(names(tbam), "_LoFreq")))
paramList <- list(ref = ref,
                  region = "apps/variants/vcf/intval.bed",
                  dbsnp = dbsnp,
                  threads = 16)

res11 <- runCWLBatch(LoFreq, outdir = "apps/variants/LoFreq", inputList, paramList,
                     BPPARAM = BatchtoolsParam(workers = 2,
                                               cluster = "sge",
                                               template = "apps/data/sge.tmpl",
                                               resources = list(threads = 16,
                                                                queue = "all.q",
                                                                jobname = "LoFreq"),
                                               log = TRUE, logdir = ".", progressbar = TRUE),
                     stderr = "")
```

Here are the results:
```{r}
dir_info("apps/variants/LoFreq", glob = "*.vcf.gz", recurse = TRUE) %>% select(path, size)
```

More details can be found:
<https://csb5.github.io/lofreq/>

### VarScan2
We simply prepare inputs and run as before.
```{r, eval=FALSE}
inputList <- list(tbam = tbam,
                  nbam = nbam)
paramList <- list(ref = ref,
                  region = "apps/variants/vcf/intval.bed")

res13 <- runCWLBatch(VarScan2Somatic, outdir = "apps/variants/VarScan2", inputList, paramList,
                     BPPARAM = BatchtoolsParam(workers = 2,
                                               cluster = "sge",
                                               template = "apps/data/sge.tmpl",
                                               resources = list(threads = 16,
                                                                queue = "all.q",
                                                                jobname = "VarScan2"),
                                               log = TRUE, logdir = ".", progressbar = TRUE),
                     stderr = "")
```

Here are the results:
```{r}
dir_info("apps/variants/VarScan2", glob = "*.vcf", recurse = TRUE) %>% select(path, size)
```

More details can be found:
<http://varscan.sourceforge.net/>

### Lancet
The tool requires BAM files, reference genome and capture region.
```{r, eval=FALSE}
inputList <- list(tbam = tbam,
                  nbam = nbam)
paramList <- list(ref = ref,
                  region = "apps/variants/vcf/intval.bed",
                  threads = 16)

res14 <- runCWLBatch(lancet, outdir = "apps/variants/lancet", inputList, paramList,
                     BPPARAM = BatchtoolsParam(workers = 2,
                                               cluster = "sge",
                                               template = "apps/data/sge.tmpl",
                                               resources = list(threads = 16,
                                                                queue = "all.q",
                                                                jobname = "lancet"),
                                               log = TRUE, logdir = ".", progressbar = TRUE),
                     stderr = "")

```

Here are the results:
```{r}
dir_info("apps/variants/lancet", glob = "*.vcf", recurse = TRUE) %>% select(path, size)
```

More details can be found:
<https://github.com/nygenome/lancet>


### Neusomatic
It is recommended to combine the mutation results from multiple somatic mutation callers to increase both sensitivity and specificitiy by recent publications. The recent published tools `neusomatic` provides a emsembling mode to combine multiple existing methods to achieve the highest accuracy.

First, we collect the outputs from previous somatic variant callers.
```{r, eval=FALSE}
mutect2Vcf <- list.files("apps/variants/mutect2", "obfiltered.vcf$",
                         recursive = TRUE, full.names = TRUE)
varscan2Snp <- list.files("apps/variants/VarScan2", ".snp.vcf$",
                          recursive = TRUE, full.names = TRUE)
varscan2Indel <- list.files("apps/variants/VarScan2", ".indel.vcf$",
                          recursive = TRUE, full.names = TRUE)
sniperVcf <- list.files("apps/variants/SomaticSniper", "vcf$",
                        recursive = TRUE, full.names = TRUE)
vardictVcf <- list.files("apps/variants/VarDict", ".vcf$",
                         recursive = TRUE, full.names = TRUE)
museVcf <- list.files("apps/variants/MuSE", ".vcf$",
                      recursive = TRUE, full.names = TRUE)
strelkaSNP <- list.files("apps/variants/strelka", "snvs.vcf.gz$",
                         recursive = TRUE, full.names = TRUE)
strelkaIndel <- list.files("apps/variants/strelka", "indels.vcf.gz$",
                           recursive = TRUE, full.names = TRUE)
lofreqSNP <- list.files("apps/variants/LoFreq", "final.snvs.vcf.gz$",
                        recursive = TRUE, full.names = TRUE)
lofreqIndel <- list.files("apps/variants/LoFreq", "final.indels.vcf.gz$",
                          recursive = TRUE, full.names = TRUE)
```

Here we use the wrapper script from somaticseq to essemble the results.
```{r, eval=FALSE}
inputList <- list(tbam = tbam,
                  nbam = nbam,
                  mutect2 = as.list(mutect2Vcf),
                  varscanSnv = as.list(varscan2Snp),
                  varscanIndel = as.list(varscan2Indel),
                  sniper = as.list(sniperVcf),
                  vardict = as.list(vardictVcf),
                  muse = as.list(museVcf),
                  strelkaSnv = as.list(strelkaSNP),
                  strelkaIndel = as.list(strelkaIndel),
                  lofreqSnv = as.list(lofreqSNP),
                  lofreqIndel = as.list(lofreqIndel))
paramList <- list(ref = ref,
                  region = "apps/variants/vcf/intval.bed",
                  dbsnp = dbsnp)

res15 <- runCWLBatch(SomaticSeq_Wrapper, outdir = "apps/variants/Wrapper", inputList, paramList,
                     BPPARAM = BatchtoolsParam(workers = 2,
                                               cluster = "sge",
                                               template = "apps/data/sge.tmpl",
                                               resources = list(threads = 16,
                                                                queue = "mem.q",
                                                                jobname = "Wrapper"),
                                               log = TRUE, logdir = ".", progressbar = TRUE),
                     stderr = "")
```

The `neusomatic` pipeline provided a ensemble mode to use outputs from different somatic callers as candidates and applied a pre-trained deep learning model to filter the variant candiates.

First, we need to combine the SNV and INDEL candidates for the ensemble mode input.
```{r, eval=FALSE}
for(id in names(tbam)) {
    snv1 <- read.table(paste0("apps/variants/Wrapper/", id, "/Ensemble.sSNV.tsv"), header = TRUE)
    indel1 <- read.table(paste0("apps/variants/Wrapper/", id, "/Ensemble.sINDEL.tsv"), header = TRUE)
    var1 <- rbind(snv1, indel1)
    var1[is.na(var1)] <- 0
    write.table(var1, file = paste0("apps/variants/Wrapper/", id, "/Ensemble.sVar.tsv"),
                row.names = FALSE, sep = "\t", quote = FALSE)
}
ensemble <- list.files("apps/variants/Wrapper", "Ensemble.sVar.tsv",
                       recursive = TRUE, full.names = TRUE)
```

Then we can prepare the inputs and run the `neusomatic` pipeline.
```{r, eval=FALSE}
inputList <- list(tbam = tbam,
                  nbam = nbam,
                  ensemble = as.list(ensemble),
                  ovcf = as.list(paste0(names(tbam), "_neusomtic.vcf")))
paramList <- list(ref = ref,
                  region = "apps/variants/vcf/intval.bed",
                  threads = 16)

res16 <- runCWLBatch(neusomatic, outdir = "apps/variants/neusomatic", inputList, paramList,
                     BPPARAM = BatchtoolsParam(workers = 2,
                                               cluster = "sge",
                                               template = "apps/data/sge.tmpl",
                                               resources = list(threads = 16,
                                                                queue = "mem.q",
                                                                jobname = "neusomatic"),
                                               log = TRUE, logdir = ".", progressbar = TRUE),
                     stderr = "")

```

That's it! Here are the finnal results:
```{r}
dir_info("apps/variants/neusomatic", glob = "*.vcf", recurse = TRUE) %>% select(path, size)
```

### Combined somatic caller pipeline
We combined the previous multiple callers into a ensemble pipeline `SomaticCallers`. The detailed inputs, outputs and steps can be visualized by `plotCWL`.
```{r}
plotCWL(SomaticCallers)
```

The ensemble callers can be performed in parallel by samples. We can use the same inputs to run all the tools in one batch.
```{r, eval=FALSE}
inputList <- list(tbam = tbam,
                  nbam = nbam)
paramList <- list(Ref = ref,
                  interval = "apps/variants/vcf/intval.bed",
                  dbsnp = dbsnp,
                  gresource = "apps/data/Mutect2/af-only-gnomad.raw.sites.b37.vcf",
                  comvcf = "apps/data/Mutect2/GetPileupSummaries/small_exac_common_3_b37.vcf",
                  pon = "apps/variants/mutect2/pon.vcf",
                  threads = 16)

res17 <- runCWLBatch(SomaticCallers, outdir = "apps/variants/combined", inputList, paramList,
                     BPPARAM = BatchtoolsParam(workers = 2,
                                               cluster = "sge",
                                               template = "apps/data/sge.tmpl",
                                               resources = list(threads = 16,
                                                                queue = "mem.q",
                                                                jobname = "combined"),
                                               log = TRUE, logdir = ".", progressbar = TRUE),
                     stderr = "")
```

Here are the final results:
```{r}
list.files("apps/variants/combined/pt19")
```

## Structural variation
### Structural variation Engine


## Copy number variation
### cnvkit
