
# Neoantigen prediction
The "pVACseq" toolkit is a comprehensive cancer immunotherapy pipeline for identification of personalized variant antigens by integrating with DNA mutations and mRNA expression data to filter and rank candicate neoepitopes. The "POLYSOLVER" or "Optitype" is a tool for HLA typing. It can be used to limit the list of alleles.

The full pipeline is shown as follows.
```{r, echo=FALSE}
library(DiagrammeR)
mermaid("
graph LR
  A(tumor DNA reads)-->D{Alignment}
  B(normal DNA reads)-->D{Alignment}
  C(tumor RNA reads)-->E{Quantification}
  D{Alignment}-->F((tumor BAM))
  D{Alignment}-->G((normal BAM))
  F((tumor BAM))-->H{Somatic variant calling}
  G((normal BAM))-->H{Somatic variant calling}
  G((normal BAM))-->I{Germline variant calling}
  H{Somatic variant calling}-->J((Somatic VCF))
  I{Germline variant calling}-->K((Germline VCF))
  J((Somatic VCF))-->L{Annotating and Phasing}
  K((Germline VCF))-->L{Annotating and Phasing}
  E{Quantification}-->L{Annotating and Phasing}
  L{Annotation and Phasing}-->M((annotated VCF))
  L{Annotation and Phasing}-->N((phased VCF))
  F((tumor BAM))-->O{HLA typing}
  M((annotated VCF))-->P{Neoantigen prediction}
  N((phased VCF))-->P{Neoantigen prediction}
  O{HLA typing}-->P{Neoantigen prediction}
  P{Neoantigen prediction}-->Q((Ranked neoantigens))
")
```

Three major steps are built to run the full pipeline, including DNA variant calling, variant annotation and neoantigen prediction.

* DNA variant calling: somatic and germline variant calling
* Variant annotation: Adding DNA and RNA allele frequencies, transcript and gene level expressions.
* Neoantigen prediction: HLA typing, neoantigen filtering and ranking

First of all, load the required packages.
```{r}
library(RcwlPipelines)
library(BiocParallel)
suppressPackageStartupMessages(library(VariantAnnotation))
library(fs)
library(dplyr)
library(conflicted)
## always use select from dplyr
conflict_prefer("select", "dplyr", quiet = TRUE)
```

## Input file preparation

### Annotating VCF
First, we need to annotate somatic mutations using the tool "vep" with "Downstream" and "Wildtype" plugins. A cache of annotation sources is recommended to use "VEP" in the fastest and most efficient way. <http://useast.ensembl.org/info/docs/tools/vep/script/vep_cache.html>
```{r, eval=FALSE}
ref <- "apps/data/hs37d5.fa"
vcfs <- as.list(list.files("apps/variants/combined", ".neusomatic.vcf",
                           recursive = TRUE, full.names = TRUE))
names(vcfs) <- c("pt19", "pt20")
ovcfs <- as.list(paste0(names(vcfs), "_vep.vcf"))
inputList <- list(ivcf = vcfs,
                  ovcf = ovcfs)
paramList <- list(ref = ref,
                  cacheDir = "~/.vep")

res1 <- runCWLBatch(vep, outdir = "apps/variants/neoantigen", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "vep"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")

```

Here are the annotated VCFs.
```{r}
dir_info("apps/variants/neoantigen", recurse = TRUE, glob = "*vep.vcf") %>% dplyr::select(path, size)
```

### Adding coverage
This step use `bam-readcount` to to retrieve the number of reads in given VCF positions. Then `vcf-readcount-annotator` is used to add the read counts for snvs and indels to the input VCF file.
```{r}
plotCWL(vcfCoverage)
```

Let's prepare the inputs for DNA samples and run the `vcfCoverage` pipeline.
```{r, eval=FALSE}
vcfs <- as.list(list.files("apps/variants/neoantigen", "*vep.vcf",
                           recursive = TRUE, full.names = TRUE))
bams <- as.list(list.files("apps/variants/BAM", "tumor.*.bam$",
                           recursive = TRUE, full.names = TRUE))
samples  <-  as.list(c("SAMPLE", "SAMPLE"))
names(vcfs) <- names(bams) <- names(samples) <- c("pt19", "pt20")

inputList <- list(vcf = vcfs,
                  sample = samples,
                  bam = bams)
paramList <- list(ref = ref)

res2 <- runCWLBatch(vcfCoverage, outdir = "apps/variants/neoantigen", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "coverage"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")

```

The Allelic depths, "AD", for the ref and alt alleles were added to the VCF file.
```{r}
vcf1 <- readVcf("apps/variants/neoantigen/pt19/pt19_vep_dc_snv_indel.vcf")
geno(vcf1)
do.call(rbind, head(geno(vcf1)$AD))
```

We also could run the pipeline for RNA samples with "ntype='RNA'". Here we don't have their RNA BAM files, so the DNA BAM files are used for demonstration.
```{r, eval=FALSE}
vcfs <- as.list(list.files("apps/variants/neoantigen", "*vep_dc_snv_indel.vcf",
                           recursive = TRUE, full.names = TRUE))
bams <- as.list(list.files("apps/variants/BAM", "tumor.*.bam$",
                           recursive = TRUE, full.names = TRUE))
samples  <-  as.list(c("SAMPLE", "SAMPLE"))
names(vcfs) <- names(bams) <- names(samples) <- c("pt19", "pt20")

inputList <- list(vcf = vcfs,
                  sample = samples,
                  bam = bams)
paramList <- list(ref = ref,
                  ntype = "RNA")

res2a <- runCWLBatch(vcfCoverage, outdir = "apps/variants/neoantigen", inputList, paramList,
                     BPPARAM = BatchtoolsParam(workers = 2,
                                               cluster = "sge",
                                               template = "apps/data/sge.tmpl",
                                               resources = list(threads = 16,
                                                                queue = "all.q",
                                                                jobname = "coverage"),
                                               log = TRUE, logdir = ".", progressbar = TRUE),
                     stderr = "")
```

The Allelic depths, "RAD", for the ref and alt alleles were added to the VCF file.
```{r}
vcf1a <- readVcf("apps/variants/neoantigen/pt19/pt19_vep_dc_snv_indel_dc_snv_indel.vcf")
geno(vcf1a)
```

### Adding expression
For demostration purpose, we use the same DNA fastq files as RNASeq inputs for transcriptome quantification with `kallisto_quant`.
```{r, eval=FALSE}
fqs <- list.files("apps/variants/fastq", recursive = TRUE, full.names = TRUE)
fqs <- fqs[grep("tumor", fqs)]
fqL <- tapply(fqs, rep(c("pt19", "pt20"), each = 2), as.list)

inputList <- list(fastq = fqL)
paramList <- list(index = "apps/data/gencode.v25.transcripts_kallisto",
                  threads = 16)

res3 <- runCWLBatch(kallisto_quant, outdir = "apps/variants/neoantigen", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = length(samples),
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "kallisto"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")
```

The tool `vcf_expression_annotator` can be used to parse gene or transcript expression results and add to a given VCF. 

The abundance tables from `kallisto quant` have full transcript IDs with gene IDs annotation in the "target_id" column. We need to clean the "targe_id" column by removing the annotation parts other than "transcript_id".
```{r, eval=FALSE}
cleanExp <- function(afile) {
    exp1 <- read.table(afile, header = TRUE, stringsAsFactors = FALSE)
    exp1[,1] <- sub("\\|ENSG.*", "", exp1[,1])
    write.table(exp1, file = "abundance_clean.tsv",
                row.names = FALSE, quote = FALSE, sep = "\t")
}
p1 <- InputParam(id = "afile", type = "File",
                 prefix = "afile=", separate = FALSE)
o1 <- OutputParam(id = "aout", type = "File", glob = "abundance_clean.tsv")
CleanExp <- cwlParam(baseCommand = cleanExp,
                     inputs = InputParamList(p1),
                     outputs = OutputParamList(o1))

exps <- as.list(list.files("apps/variants/neoantigen", "abundance.tsv",
                           recursive = TRUE, full.names = TRUE))
names(exps) <- c("pt19", "pt20")

inputList <- list(afile = exps)
res4 <- runCWLBatch(CleanExp, outdir = "apps/variants/neoantigen", inputList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "vep"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")

```

Then the cleaned abundance file can be used to add expression values to the VEP annotated VCF file using the tool `vcf_expression_annotator`.
```{r, eval=FALSE}
vcfs <- as.list(list.files("apps/variants/neoantigen", "*indel_dc_snv_indel.vcf",
                           recursive = TRUE, full.names = TRUE))
exps <- as.list(list.files("apps/variants/neoantigen", "abundance_clean.tsv",
                           recursive = TRUE, full.names = TRUE))
ovcf <- list("pt19_Ann.vcf", "pt20_Ann.vcf")
names(vcfs) <- names(exps) <- names(ovcf) <- c("pt19", "pt20")

inputList <- list(ivcf = vcfs,
                  expression = exps,
                  ovcf = ovcf)
paramList <- list(etype = "kallisto",
                  gtype = "transcript")

res5 <- runCWLBatch(vcf_expression_annotator, outdir = "apps/variants/neoantigen", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "vep"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")
```

The transcript expression values, "TX", were added to the VCF file.
```{r}
vcf2 <- readVcf("apps/variants/neoantigen/pt19/pt19_Ann.vcf")
geno(vcf2)
geno(vcf2)$TX[[1]]
```

The gene level expression 
```{r, eval=FALSE}
library(tximport)

t2gene <- function(kexp){
    e1 <- read.table(kexp, header = TRUE, check.names = FALSE,
                     stringsAsFactors = FALSE, sep = "\t")
    ids <- do.call(rbind, base::strsplit(e1$target_id, split = "\\|"))
    tx2gene  <- data.frame(ids[,1:2])
    gexp <- tximport::tximport(kexp, type = "kallisto", tx2gene = tx2gene, ignoreAfterBar=TRUE)
    gExp <- data.frame(gene = sub("\\..*", "", rownames(gexp$abundance)),
                       abundance = gexp$abundance)
    write.table(gExp, file = "abundance_gene.tsv", row.names = FALSE,
                col.names = TRUE, quote = FALSE, sep = "\t")
}
p1 <- InputParam(id = "kexp", type = "File",
                 prefix = "kexp=", separate = FALSE)
o1 <- OutputParam(id = "gout", type = "File", glob = "abundance_gene.tsv")
T2Gene <- cwlParam(baseCommand = t2gene,
                   inputs = InputParamList(p1),
                   outputs = OutputParamList(o1))
## run
exps <- as.list(list.files("apps/variants/neoantigen", "abundance.tsv",
                           recursive = TRUE, full.names = TRUE))
names(exps) <- c("pt19", "pt20")

inputList <- list(kexp = exps)
res4g <- runCWLBatch(T2Gene, outdir = "apps/variants/neoantigen", inputList,
                     BPPARAM = BatchtoolsParam(workers = 2,
                                               cluster = "sge",
                                               template = "apps/data/sge.tmpl",
                                               resources = list(threads = 16,
                                                                queue = "all.q",
                                                                jobname = "vep"),
                                               log = TRUE, logdir = ".", progressbar = TRUE),
                     stderr = "")
```

Let's annotate the VCFs with gene level expression values.
```{r, eval=FALSE}
vcfs <- as.list(list.files("apps/variants/neoantigen", "_Ann.vcf$",
                           recursive = TRUE, full.names = TRUE))
exps <- as.list(list.files("apps/variants/neoantigen", "abundance_gene.tsv",
                           recursive = TRUE, full.names = TRUE))
ovcf <- list("pt19_ANN.vcf", "pt20_ANN.vcf")
names(vcfs) <- names(exps) <- names(ovcf) <- c("pt19", "pt20")

inputList <- list(ivcf = vcfs,
                  expression = exps,
                  ovcf = ovcf)
paramList <- list(etype = "custom",
                  gtype = "gene",
                  idCol = "gene",
                  expCol = "abundance")

res5g <- runCWLBatch(vcf_expression_annotator, outdir = "apps/variants/neoantigen", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "vep"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")

```

The steps were also combined into an expression annotation pipelines, `vcfExpression`. By given RNASeq raw reads, the transcript expression levels will be added to the corresponding variants in the "vep" annotated VCFs.
```{r}
plotCWL(vcfExpression)
```

### Creating a phased VCF
The germline variants and somatic variants were combined and phased in this step to calculated wildtype and mutant protein sequences more accurately. The `picard` and `GATK` tools were used to rename, combine, sort and phase the variants.

```{r}
plotCWL(phaseVcf)
```

To create a phased VCF.
```{r, eval=FALSE}
svcfs <- list(pt19 = "apps/variants/neoantigen/pt19/pt19_Ann.vcf.gz",
              pt20 = "apps/variants/neoantigen/pt20/pt20_Ann.vcf.gz")
gvcfs <- list(pt19 = "apps/variants/vcf/res/output/germline.vcf.gz",
              pt20 = "apps/variants/vcf/res/output/germline.vcf.gz")
bam <- list(pt19 = "apps/variants/BAM/tumor19/tumor19.bam",
            pt20 = "apps/variants/BAM/tumor20/tumor20.bam")
outvcf <- list(pt19 = "pt19_phased.vcf",
               pt20 = "pt20_phased.vcf")
nsamples <- list("normal19", "normal20")
tsamples <- list("tumor19", "tumor20")
inputList <- list(gvariant = gvcfs,
                  svariant = svcfs,
                  bam = bam,
                  outvcf = outvcf,
                  nsample = nsamples,
                  tsample = tsamples)
paramList <- list(ref = ref)

res6 <- runCWLBatch(phaseVcf, outdir = "apps/variants/neoantigen", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "phase"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")

```

The phased VCF file using HP tags to link alleles. Here are the phased VCFs.
```{r}
dir_info("apps/variants/neoantigen", recurse = TRUE, glob = "*phased.vcf*") %>% dplyr::select(path, size)
```

### Combined VCF preparing pipeline
The previous 4 annotation steps have been intergrated into one VCF annotation pipeline. Thus all the required inputs can be prepared in just one step.
```{r}
plotCWL(AnnPhaseVcf)
```

Here are the example to run the annotation and phasing pipeline.
```{r, eval=FALSE}
svcfs <- as.list(list.files("apps/variants/combined", ".neusomatic.vcf",
                           recursive = TRUE, full.names = TRUE))
names(svcfs) <- c("pt19", "pt20")
gvcfs <- list(pt19 = "apps/variants/vcf/res/output/germline.vcf.gz",
              pt20 = "apps/variants/vcf/res/output/germline.vcf.gz")
bam <- list(pt19 = "apps/variants/BAM/tumor19/tumor19.bam",
            pt20 = "apps/variants/BAM/tumor20/tumor20.bam")
## use rnaseq BAM instead
rbam <- list(pt19 = "apps/variants/BAM/tumor19/tumor19.bam",
             pt20 = "apps/variants/BAM/tumor20/tumor20.bam")
fqs <- list.files("apps/variants/fastq", recursive = TRUE, full.names = TRUE)
fqs <- fqs[grep("tumor", fqs)]
fqL <- tapply(fqs, rep(c("pt19", "pt20"), each = 2), as.list)
nsamples <- list("normal19", "normal20")
tsamples <- list("tumor19", "tumor20")

inputList <- list(gvcf = gvcfs,
                  svcf = svcfs,
                  tbam = bam,
                  nsample = nsamples,
                  tsample = tsamples,
                  rnaseqs = fqL,
                  rbam = rbam)
paramList <- list(ref = ref,
                  VepDir = "~/.vep/",
                  kallistoIdx = "apps/data/gencode.v25.transcripts_kallisto",
                  threads = 16)

res7 <- runCWLBatch(AnnPhaseVcf, outdir = "apps/variants/neoantigen", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "AnnPhase"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")

```

Here are the results:
```{r}
list.files("apps/variants/neoantigen/pt19", "vcf.gz")
```


## HLA typing
### polysolver
The `polysolver` is a tool for HLA typing based on whole exome sequencing data. The details to run the tool can be found: <https://software.broadinstitute.org/cancer/cga/polysolver_run>.
```{r, eval=FALSE}
inputList <- list(bam = bam)
res8 <- runCWLBatch(polysolver, outdir = "apps/variants/neoantigen", inputList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "polysolver"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "", cwlTemp = TRUE)

```

### Optitype

## Neoantigen prediction by pVACseq
The `pvacseq` tool take the annotated somatic variants and phased combined variants as inputs. Multiple candidate alleles from hla typing and multiple epitope prediction algorithms can be used.

```{r, eval=FALSE}
vcfs <- as.list(list.files("apps/variants/neoantigen", "gAnn.vcf.gz$",
                           recursive = TRUE, full.names = TRUE))
pvcfs <- as.list(list.files("apps/variants/neoantigen", "phased.vcf.gz$",
                           recursive = TRUE, full.names = TRUE))
names(vcfs) <- names(pvcfs) <- c("pt19", "pt20")
alleles <- list(pt19 = list("HLA-A*02:01", "HLA-B*35:01"),
                pt20 = list("HLA-A*02:01", "HLA-B*35:01", "DRB1*11:01"))

inputList <- list(ivcf = vcfs,
                  phasedVcf = pvcfs,
                  sample = tsamples,
                  allele = alleles)
paramList <- list(algorithms = list("MHCflurry", "MHCnuggetsI",
                                    "MHCnuggetsII", "NNalign", "NetMHC"))

res9 <- runCWLBatch(pvacseq, outdir = "apps/variants/neoantigen", inputList, paramList,
                    BPPARAM = BatchtoolsParam(workers = 2,
                                              cluster = "sge",
                                              template = "apps/data/sge.tmpl",
                                              resources = list(threads = 16,
                                                               queue = "all.q",
                                                               jobname = "pvacseq"),
                                              log = TRUE, logdir = ".", progressbar = TRUE),
                    stderr = "")

```

Here are the filtered and ranked results:
```{r}
list.files("apps/variants/neoantigen/pt20/pvacseq_out/combined/")
neo <- read.table("apps/variants/neoantigen/pt20/pvacseq_out/combined/tumor20.filtered.condensed.ranked.tsv", header = TRUE, sep = "\t", check.names = FALSE)
head(neo)
```
