<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 9 Bulk RNAseq | Bioinformatics tools and pipelines using R and CWL</title>
  <meta name="description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 9 Bulk RNAseq | Bioinformatics tools and pipelines using R and CWL" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 9 Bulk RNAseq | Bioinformatics tools and pipelines using R and CWL" />
  
  <meta name="twitter:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  

<meta name="author" content="Qiang Hu" />
<meta name="author" content="Qian Liu" />


<meta name="date" content="2021-02-27" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="dnaseq-variant-calling.html"/>
<link rel="next" href="single-cell-rnaseq.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/viz-1.8.2/viz.js"></script>
<link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="libs/grViz-binding-1.0.6.1/grViz.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Rcwl and Case Study</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#r-package-installation"><i class="fa fa-check"></i><b>0.1</b> R package installation</a></li>
<li class="chapter" data-level="0.2" data-path="index.html"><a href="index.html#system-requirements"><i class="fa fa-check"></i><b>0.2</b> System requirements</a></li>
<li class="chapter" data-level="0.3" data-path="index.html"><a href="index.html#docker"><i class="fa fa-check"></i><b>0.3</b> Docker</a></li>
<li class="chapter" data-level="0.4" data-path="index.html"><a href="index.html#structure-of-the-book"><i class="fa fa-check"></i><b>0.4</b> Structure of the book</a></li>
<li class="chapter" data-level="0.5" data-path="index.html"><a href="index.html#r-session-information"><i class="fa fa-check"></i><b>0.5</b> R session information</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="get-started.html"><a href="get-started.html"><i class="fa fa-check"></i><b>2</b> Get started</a></li>
<li class="chapter" data-level="3" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html"><i class="fa fa-check"></i><b>3</b> Wrap command line tools</a>
<ul>
<li class="chapter" data-level="3.1" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#input-parameters"><i class="fa fa-check"></i><b>3.1</b> Input Parameters</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#essential-input-parameters"><i class="fa fa-check"></i><b>3.1.1</b> Essential Input parameters</a></li>
<li class="chapter" data-level="3.1.2" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#array-inputs"><i class="fa fa-check"></i><b>3.1.2</b> Array Inputs</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#output-parameters"><i class="fa fa-check"></i><b>3.2</b> Output Parameters</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#capturing-output"><i class="fa fa-check"></i><b>3.2.1</b> Capturing Output</a></li>
<li class="chapter" data-level="3.2.2" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#array-outputs"><i class="fa fa-check"></i><b>3.2.2</b> Array Outputs</a></li>
<li class="chapter" data-level="3.2.3" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#standard-output"><i class="fa fa-check"></i><b>3.2.3</b> Standard output</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="writing-pipeline.html"><a href="writing-pipeline.html"><i class="fa fa-check"></i><b>4</b> Writing Pipeline</a>
<ul>
<li class="chapter" data-level="4.1" data-path="writing-pipeline.html"><a href="writing-pipeline.html#scattering-pipeline"><i class="fa fa-check"></i><b>4.1</b> Scattering pipeline</a></li>
<li class="chapter" data-level="4.2" data-path="writing-pipeline.html"><a href="writing-pipeline.html#pipeline-plot"><i class="fa fa-check"></i><b>4.2</b> Pipeline plot</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html"><i class="fa fa-check"></i><b>5</b> Tool/pipeline execution</a>
<ul>
<li class="chapter" data-level="5.1" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#running-tools-in-docker"><i class="fa fa-check"></i><b>5.1</b> Running Tools in Docker</a></li>
<li class="chapter" data-level="5.2" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#running-tools-in-cluster-server"><i class="fa fa-check"></i><b>5.2</b> Running Tools in Cluster server</a></li>
<li class="chapter" data-level="5.3" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#web-application"><i class="fa fa-check"></i><b>5.3</b> Web Application</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#cwlprocess-example"><i class="fa fa-check"></i><b>5.3.1</b> cwlProcess example</a></li>
<li class="chapter" data-level="5.3.2" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#cwlprocess-to-shiny-app"><i class="fa fa-check"></i><b>5.3.2</b> cwlProcess to Shiny App</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html"><i class="fa fa-check"></i><b>6</b> RcwlPipelines</a>
<ul>
<li class="chapter" data-level="6.1" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#rcwl-recipes-and-cwl-scripts"><i class="fa fa-check"></i><b>6.1</b> Rcwl recipes and CWL scripts</a></li>
<li class="chapter" data-level="6.2" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#rcwlpipelines-core-functions"><i class="fa fa-check"></i><b>6.2</b> RcwlPipelines core functions</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#cwlupdate"><i class="fa fa-check"></i><b>6.2.1</b> <code>cwlUpdate</code></a></li>
<li class="chapter" data-level="6.2.2" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#cwlsearch"><i class="fa fa-check"></i><b>6.2.2</b> <code>cwlSearch</code></a></li>
<li class="chapter" data-level="6.2.3" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#cwlload"><i class="fa fa-check"></i><b>6.2.3</b> <code>cwlLoad</code></a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#toolpipeline-customization"><i class="fa fa-check"></i><b>6.3</b> Tool/pipeline customization</a></li>
<li class="chapter" data-level="6.4" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#build-a-pipeline"><i class="fa fa-check"></i><b>6.4</b> Build a pipeline</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="dnaseq-alignment.html"><a href="dnaseq-alignment.html"><i class="fa fa-check"></i><b>7</b> DNAseq alignment</a></li>
<li class="chapter" data-level="8" data-path="dnaseq-variant-calling.html"><a href="dnaseq-variant-calling.html"><i class="fa fa-check"></i><b>8</b> DNAseq variant calling</a></li>
<li class="chapter" data-level="9" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html"><i class="fa fa-check"></i><b>9</b> Bulk RNAseq</a>
<ul>
<li class="chapter" data-level="9.1" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#prepare-data"><i class="fa fa-check"></i><b>9.1</b> Prepare data</a></li>
<li class="chapter" data-level="9.2" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#submit-parallel-jobs"><i class="fa fa-check"></i><b>9.2</b> Submit parallel jobs</a></li>
<li class="chapter" data-level="9.3" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#qc-summary"><i class="fa fa-check"></i><b>9.3</b> QC Summary</a></li>
<li class="chapter" data-level="9.4" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#abundances-summary"><i class="fa fa-check"></i><b>9.4</b> Abundances summary</a></li>
<li class="chapter" data-level="9.5" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#transcriptome-quantification"><i class="fa fa-check"></i><b>9.5</b> transcriptome quantification</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#kallisto"><i class="fa fa-check"></i><b>9.5.1</b> Kallisto</a></li>
<li class="chapter" data-level="9.5.2" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#salmon"><i class="fa fa-check"></i><b>9.5.2</b> salmon</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="single-cell-rnaseq.html"><a href="single-cell-rnaseq.html"><i class="fa fa-check"></i><b>10</b> single cell RNAseq</a></li>
<li class="chapter" data-level="11" data-path="mirna.html"><a href="mirna.html"><i class="fa fa-check"></i><b>11</b> miRNA</a>
<ul>
<li class="chapter" data-level="11.0.1" data-path="mirna.html"><a href="mirna.html#reference-index"><i class="fa fa-check"></i><b>11.0.1</b> Reference index</a></li>
<li class="chapter" data-level="11.0.2" data-path="mirna.html"><a href="mirna.html#run-mirdeep2-pipeline"><i class="fa fa-check"></i><b>11.0.2</b> Run miRDeep2 pipeline</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics tools and pipelines using R and CWL</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="bulk-rnaseq" class="section level1" number="9">
<h1><span class="header-section-number">Chapter 9</span> Bulk RNAseq</h1>
<p>This pipeline covers RNA-seq reads quality summary by <code>fastQC</code>,
alignment by <code>STAR</code>, quantification by <code>featureCounts</code> and quality
control by <code>RSeQC</code>.</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="bulk-rnaseq.html#cb160-1"></a>rnaseq_Sf &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="st">&quot;pl_rnaseq_Sf&quot;</span>)</span></code></pre></div>
<pre><code>## fastqc loaded</code></pre>
<pre><code>## STAR loaded</code></pre>
<pre><code>## sortBam loaded</code></pre>
<pre><code>## samtools_index loaded</code></pre>
<pre><code>## samtools_flagstat loaded</code></pre>
<pre><code>## featureCounts loaded</code></pre>
<pre><code>## gtfToGenePred loaded</code></pre>
<pre><code>## genePredToBed loaded</code></pre>
<pre><code>## read_distribution loaded</code></pre>
<pre><code>## geneBody_coverage loaded</code></pre>
<pre><code>## gCoverage loaded</code></pre>
<pre><code>## STAR loaded</code></pre>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="bulk-rnaseq.html#cb173-1"></a><span class="kw">plotCWL</span>(rnaseq_Sf)</span></code></pre></div>
<div id="htmlwidget-f61a4cefaa6c7cbbde4e" style="width:672px;height:480px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-f61a4cefaa6c7cbbde4e">{"x":{"diagram":"digraph {\n\ngraph [layout = \"neato\",\n       outputorder = \"edgesfirst\",\n       bgcolor = \"white\"]\n\nnode [fontname = \"Helvetica\",\n      fontsize = \"10\",\n      shape = \"circle\",\n      fixedsize = \"true\",\n      width = \"0.5\",\n      style = \"filled\",\n      fillcolor = \"aliceblue\",\n      color = \"gray70\",\n      fontcolor = \"gray50\"]\n\nedge [fontname = \"Helvetica\",\n     fontsize = \"8\",\n     len = \"1.5\",\n     color = \"gray80\",\n     arrowsize = \"0.5\"]\n\n  \"1\" [label = \"in_seqfiles\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"-1,7!\"] \n  \"2\" [label = \"in_prefix\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"0,7!\"] \n  \"3\" [label = \"in_genomeDir\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1.5,7!\"] \n  \"4\" [label = \"in_GTFfile\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"3.5,7!\"] \n  \"5\" [label = \"in_runThreadN\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2.5,7!\"] \n  \"6\" [label = \"out_fastqc\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"-1,5!\"] \n  \"7\" [label = \"out_BAM\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1,3!\"] \n  \"8\" [label = \"out_Log\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"0,5!\"] \n  \"9\" [label = \"out_Count\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"4,5!\"] \n  \"10\" [label = \"out_stat\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,3!\"] \n  \"11\" [label = \"out_count\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"5.5,1!\"] \n  \"12\" [label = \"out_distribution\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"0,1!\"] \n  \"13\" [label = \"out_gCovP\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1,1!\"] \n  \"14\" [label = \"out_gCovT\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2.5,1!\"] \n  \"15\" [label = \"fastqc\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"-1,6!\"] \n  \"16\" [label = \"STAR\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1.5,6!\"] \n  \"17\" [label = \"sortBam\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1.5,5!\"] \n  \"18\" [label = \"samtools_index\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1,4!\"] \n  \"19\" [label = \"samtools_flagstat\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,4!\"] \n  \"20\" [label = \"featureCounts\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"5.5,2!\"] \n  \"21\" [label = \"gtfToGenePred\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"3.5,6!\"] \n  \"22\" [label = \"genePredToBed\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"3,5!\"] \n  \"23\" [label = \"r_distribution\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"0,2!\"] \n  \"24\" [label = \"gCoverage\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,2!\"] \n  \"1\"->\"15\" \n  \"15\"->\"6\" \n  \"2\"->\"16\" \n  \"3\"->\"16\" \n  \"4\"->\"16\" \n  \"1\"->\"16\" \n  \"5\"->\"16\" \n  \"16\"->\"8\" \n  \"16\"->\"9\" \n  \"16\"->\"17\" \n  \"17\"->\"18\" \n  \"18\"->\"7\" \n  \"17\"->\"19\" \n  \"19\"->\"10\" \n  \"4\"->\"20\" \n  \"7\"->\"20\" \n  \"20\"->\"11\" \n  \"4\"->\"21\" \n  \"21\"->\"22\" \n  \"7\"->\"23\" \n  \"22\"->\"23\" \n  \"23\"->\"12\" \n  \"7\"->\"24\" \n  \"22\"->\"24\" \n  \"24\"->\"13\" \n  \"24\"->\"14\" \n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>
<p>The pipeline includes 10 steps, each step runs a single command as
follows:</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="bulk-rnaseq.html#cb174-1"></a><span class="kw">runs</span>(rnaseq_Sf)</span></code></pre></div>
<pre><code>## List of length 10
## names(10): fastqc STAR sortBam ... genePredToBed r_distribution gCoverage</code></pre>
<ul>
<li><code>fastqc</code>: to run quality summary for raw fastqs with base command <code>fastqc</code>.</li>
<li><code>STAR</code>: to align fastqs with <code>STAR</code>.</li>
<li><code>sortBam</code>: to sort bam files with <code>samtools</code>.</li>
<li><code>samtools_index</code>: to index aligned bam file with <code>samtools</code>.</li>
<li><code>samtools_flagstat</code>: to summarize alignment flags with <code>samtools</code>.</li>
<li><code>featureCounts</code>: to quantify gene abundance with <code>featureCounts</code>.</li>
<li><code>gtfToGenePred</code>: to convert GTF annotation to ‘genePred’ format with <code>RSeQC</code>.</li>
<li><code>genePredToBed</code>: to convert ‘genePred’ annotation to ‘bed’ format with <code>RSeQC</code>.</li>
<li><code>r_distribution</code>: to run read distribution over genome features with <code>RSeQC</code>.</li>
<li><code>gCoverage</code>: to summarize read coverage over gene body with <code>RSeQC</code>.</li>
</ul>
<p>The <code>rnaseq_Sf</code> pipepine output includes the QC result from <code>fastqc</code>
step, indexed bam files from <code>samtools_index</code> step, log and read
counts from <code>STAR</code> step, flag summary from <code>samtools_flagstat</code> step,
feature counts from <code>featureCounts</code> step, alignment QC results from
<code>RSeQC</code> steps.</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="bulk-rnaseq.html#cb176-1"></a><span class="kw">outputs</span>(rnaseq_Sf)</span></code></pre></div>
<pre><code>## outputs:
## out_fastqc:
##   type: File[]
##   outputSource: fastqc/QCfile
## out_BAM:
##   type: File
##   outputSource: samtools_index/idx
## out_Log:
##   type: File
##   outputSource: STAR/outLog
## out_Count:
##   type: File
##   outputSource: STAR/outCount
## out_stat:
##   type: File
##   outputSource: samtools_flagstat/flagstat
## out_count:
##   type: File
##   outputSource: featureCounts/Count
## out_distribution:
##   type: File
##   outputSource: r_distribution/distOut
## out_gCovP:
##   type: File
##   outputSource: gCoverage/gCovPDF
## out_gCovT:
##   type: File
##   outputSource: gCoverage/gCovTXT</code></pre>
<div id="prepare-data" class="section level2" number="9.1">
<h2><span class="header-section-number">9.1</span> Prepare data</h2>
<p>An RNASeq test data with paired-end fastq files for 6 samples can be
downloaded from <a href="http://genomedata.org/rnaseq-tutorial">genomedata</a>.
Create a local directory and follow the code below to download and
uncompress.</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="bulk-rnaseq.html#cb178-1"></a><span class="kw">dir.create</span>(<span class="st">&quot;data/RNAseq&quot;</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</span>
<span id="cb178-2"><a href="bulk-rnaseq.html#cb178-2"></a><span class="kw">download.file</span>(<span class="st">&quot;http://genomedata.org/rnaseq-tutorial/HBR_UHR_ERCC_ds_5pc.tar&quot;</span>,</span>
<span id="cb178-3"><a href="bulk-rnaseq.html#cb178-3"></a>              <span class="st">&quot;data/RNAseq/HBR_UHR_ERCC_ds_5pc.tar&quot;</span>)</span>
<span id="cb178-4"><a href="bulk-rnaseq.html#cb178-4"></a><span class="kw">untar</span>(<span class="st">&quot;data/RNAseq/HBR_UHR_ERCC_ds_5pc.tar&quot;</span>,</span>
<span id="cb178-5"><a href="bulk-rnaseq.html#cb178-5"></a>      <span class="dt">exdir =</span> <span class="st">&quot;data/RNAseq/&quot;</span>)</span></code></pre></div>
</div>
<div id="submit-parallel-jobs" class="section level2" number="9.2">
<h2><span class="header-section-number">9.2</span> Submit parallel jobs</h2>
<p>Powered by <code>BiocParallel</code>,<code>Rcwl</code> supports parallel job running for
multiple samples using the <code>runCWLBatch</code> function.</p>
<p>The <code>BPPARAM</code> argument in <code>runCWLBatch()</code> defines the parallel
parameters. It can be defined by <code>BiocParallel::BatchtoolsParam</code>
function, where the <code>cluster</code> argument takes different values for
different cluster job manager, such as “multicore”, “sge” and
“slurm”. More details about available options can be checked by
<code>?BiocParallel::BatchtoolsParam</code>.</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="bulk-rnaseq.html#cb179-1"></a><span class="kw">library</span>(BiocParallel)</span></code></pre></div>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="bulk-rnaseq.html#cb180-1"></a>bpparam &lt;-<span class="st"> </span><span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>, <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb180-2"><a href="bulk-rnaseq.html#cb180-2"></a>                           <span class="dt">template =</span> <span class="kw">batchtoolsTemplate</span>(<span class="st">&quot;sge&quot;</span>))</span></code></pre></div>
<p>In the following example, we are using “multicore” for the parallel
running.</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="bulk-rnaseq.html#cb181-1"></a>bpparam &lt;-<span class="st"> </span><span class="kw">BatchtoolsParam</span>(</span>
<span id="cb181-2"><a href="bulk-rnaseq.html#cb181-2"></a>    <span class="dt">workers =</span> <span class="dv">2</span>, <span class="dt">cluster =</span> <span class="st">&quot;multicore&quot;</span>)</span></code></pre></div>
<p>When submitting parallel jobs using <code>runCWLBatch</code> function, two other
arguments: <code>inputList</code> and <code>paramList</code>, need to be defined.</p>
<p>The <code>inputList</code> argument is required to be a list of input parameter
values for samples that are to be computed parallelly. <strong>NOTE</strong> that
the names of the list must be consistent with the ids of input
parameters. In this example, they are:</p>
<ul>
<li><code>in_seqfiles</code>: A list with the fastq files of each sample in each
element. The names of the list need to be defined and can be the
sample IDs. The length of the list will be the same as the number of
samples, so that the list of samples can be assigned to different
nodes for parallel computing. Here we only use 2 samples for
demonstration purposes.</li>
<li><code>in_prefix</code>: A list of sample IDs.</li>
</ul>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="bulk-rnaseq.html#cb182-1"></a>files &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="kw">list.files</span>(<span class="st">&quot;data/RNAseq/&quot;</span>, <span class="st">&quot;.gz&quot;</span>,</span>
<span id="cb182-2"><a href="bulk-rnaseq.html#cb182-2"></a>                                  <span class="dt">full.names =</span> <span class="ot">TRUE</span>))[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>]</span>
<span id="cb182-3"><a href="bulk-rnaseq.html#cb182-3"></a>files &lt;-<span class="st"> </span><span class="kw">tapply</span>(files, <span class="kw">substring</span>(<span class="kw">basename</span>(files), <span class="dv">1</span>, <span class="dv">8</span>), as.list)</span>
<span id="cb182-4"><a href="bulk-rnaseq.html#cb182-4"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">in_seqfiles =</span> files,</span>
<span id="cb182-5"><a href="bulk-rnaseq.html#cb182-5"></a>                  <span class="dt">in_prefix =</span> <span class="kw">as.list</span>(<span class="kw">names</span>(files)))</span></code></pre></div>
<p>The <code>paramList</code> argument is required to be a list of input parameter
values that are to be shared for all parallelly running samples. In
this example, they are:</p>
<ul>
<li><code>in_genomeDir</code>: The reference genome indexes for STAR.</li>
<li><code>in_GTFfile</code>: The gene annotation file in GTF format.</li>
<li><code>in_runThreadN</code>: The number of threads to run for each job.</li>
</ul>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="bulk-rnaseq.html#cb183-1"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(</span>
<span id="cb183-2"><a href="bulk-rnaseq.html#cb183-2"></a>    <span class="dt">in_genomeDir =</span> <span class="st">&quot;data/resources/GRCh38_chr22&quot;</span>,</span>
<span id="cb183-3"><a href="bulk-rnaseq.html#cb183-3"></a>    <span class="dt">in_GTFfile =</span> <span class="st">&quot;data/resources/GRCh38_chr22/gencode.v32.annotation_chr22.gtf&quot;</span>,</span>
<span id="cb183-4"><a href="bulk-rnaseq.html#cb183-4"></a>    <span class="dt">in_runThreadN =</span> <span class="dv">2</span></span>
<span id="cb183-5"><a href="bulk-rnaseq.html#cb183-5"></a>)</span></code></pre></div>
<p>Here we can also modify the default argument values in some steps of a
pipeline. For example,</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="bulk-rnaseq.html#cb184-1"></a><span class="kw">arguments</span>(rnaseq_Sf, <span class="st">&quot;STAR&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</span></code></pre></div>
<pre><code>## [[1]]
## [1] &quot;--outFilterMultimapNmax&quot;
## 
## [[2]]
## [1] &quot;3&quot;</code></pre>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="bulk-rnaseq.html#cb186-1"></a><span class="kw">arguments</span>(rnaseq_Sf, <span class="st">&quot;STAR&quot;</span>)[[<span class="dv">2</span>]] &lt;-<span class="st"> &quot;2&quot;</span></span>
<span id="cb186-2"><a href="bulk-rnaseq.html#cb186-2"></a><span class="kw">arguments</span>(rnaseq_Sf, <span class="st">&quot;STAR&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</span></code></pre></div>
<pre><code>## [[1]]
## [1] &quot;--outFilterMultimapNmax&quot;
## 
## [[2]]
## [1] &quot;2&quot;</code></pre>
<p>Now that the fastqc files of each sample will be submitted to
different nodes to run the whole pipeline parallelly.</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="bulk-rnaseq.html#cb188-1"></a>res &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(<span class="dt">cwl =</span> rnaseq_Sf,</span>
<span id="cb188-2"><a href="bulk-rnaseq.html#cb188-2"></a>                   <span class="dt">outdir =</span> <span class="st">&quot;output/RNAseq_bulk&quot;</span>,</span>
<span id="cb188-3"><a href="bulk-rnaseq.html#cb188-3"></a>                   <span class="dt">inputList =</span> inputList,</span>
<span id="cb188-4"><a href="bulk-rnaseq.html#cb188-4"></a>                   <span class="dt">paramList =</span> paramList,</span>
<span id="cb188-5"><a href="bulk-rnaseq.html#cb188-5"></a>                   <span class="dt">BPPARAM =</span> bpparam,</span>
<span id="cb188-6"><a href="bulk-rnaseq.html#cb188-6"></a>                   <span class="dt">showLog =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>Pipeline results are collected in the output directory (defined in
<code>outdir</code>) for each sample.</p>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="bulk-rnaseq.html#cb189-1"></a><span class="kw">dir</span>(<span class="st">&quot;output/RNAseq_bulk&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;HBR_Rep1&quot; &quot;HBR_Rep2&quot;</code></pre>
</div>
<div id="qc-summary" class="section level2" number="9.3">
<h2><span class="header-section-number">9.3</span> QC Summary</h2>
<p>The tool <code>multiqc</code> can aggregate results from the multiple outputs of
the pipeline and generate a single page report, which also was
implemented in the <code>RcwlPipelines</code> package:</p>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="bulk-rnaseq.html#cb191-1"></a>multiqc<span class="op">$</span>dir &lt;-<span class="st"> &quot;output/RNASeq_bulk&quot;</span></span>
<span id="cb191-2"><a href="bulk-rnaseq.html#cb191-2"></a>multiqc</span></code></pre></div>
<p>We can also run the tool using <code>Rcwl</code> locally with the option <code>docker = TRUE</code>:</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="bulk-rnaseq.html#cb192-1"></a><span class="kw">runCWL</span>(multiqc, <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>, <span class="dt">Args =</span> <span class="st">&quot;--preserve-entire-environment&quot;</span>, <span class="dt">docker =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
</div>
<div id="abundances-summary" class="section level2" number="9.4">
<h2><span class="header-section-number">9.4</span> Abundances summary</h2>
<p>Here we use the <em>R/Bioconductor</em> package <code>edgeR</code> functions to
calculate the RPKM and CPM abundances.</p>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="bulk-rnaseq.html#cb193-1"></a>countfiles &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="st">&quot;output/RNAseq_bulk&quot;</span>, <span class="st">&quot;featureCounts.txt$&quot;</span>,</span>
<span id="cb193-2"><a href="bulk-rnaseq.html#cb193-2"></a>                         <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</span>
<span id="cb193-3"><a href="bulk-rnaseq.html#cb193-3"></a>samples &lt;-<span class="st"> </span><span class="kw">basename</span>(<span class="kw">dirname</span>(countfiles))</span>
<span id="cb193-4"><a href="bulk-rnaseq.html#cb193-4"></a></span>
<span id="cb193-5"><a href="bulk-rnaseq.html#cb193-5"></a>rExp &lt;-<span class="st"> </span><span class="cf">function</span>(countfile){</span>
<span id="cb193-6"><a href="bulk-rnaseq.html#cb193-6"></a>    count1 &lt;-<span class="st"> </span><span class="kw">read.table</span>(countfile, <span class="dt">header =</span> <span class="ot">TRUE</span>)[, <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">6</span>,<span class="dv">7</span>)]</span>
<span id="cb193-7"><a href="bulk-rnaseq.html#cb193-7"></a>    rpkm1 &lt;-<span class="st"> </span>edgeR<span class="op">::</span><span class="kw">rpkm</span>(count1[,<span class="dv">3</span>,<span class="dt">drop=</span>F], <span class="dt">gene.length =</span> count1<span class="op">$</span>Length)</span>
<span id="cb193-8"><a href="bulk-rnaseq.html#cb193-8"></a>    cpm1 &lt;-<span class="st"> </span>edgeR<span class="op">::</span><span class="kw">cpm</span>(count1[,<span class="dv">3</span>])</span>
<span id="cb193-9"><a href="bulk-rnaseq.html#cb193-9"></a>    count1 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(count1, rpkm1, cpm1)</span>
<span id="cb193-10"><a href="bulk-rnaseq.html#cb193-10"></a>    <span class="kw">colnames</span>(count1)[<span class="dv">3</span><span class="op">:</span><span class="dv">5</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;count&quot;</span>, <span class="st">&quot;rpkm&quot;</span>, <span class="st">&quot;cpm&quot;</span>)</span>
<span id="cb193-11"><a href="bulk-rnaseq.html#cb193-11"></a>    <span class="kw">return</span>(count1)</span>
<span id="cb193-12"><a href="bulk-rnaseq.html#cb193-12"></a>}</span></code></pre></div>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="bulk-rnaseq.html#cb194-1"></a><span class="kw">head</span>(<span class="kw">rExp</span>(countfiles[<span class="dv">1</span>]))</span></code></pre></div>
<pre><code>##              Geneid Length count rpkm cpm
## 1 ENSG00000223972.5   1735     0    0   0
## 2 ENSG00000227232.5   1351     0    0   0
## 3 ENSG00000278267.1     68     0    0   0
## 4 ENSG00000243485.5   1021     0    0   0
## 5 ENSG00000284332.1    138     0    0   0
## 6 ENSG00000237613.2   1219     0    0   0</code></pre>
<p>We combine the files into one file, and then the data is ready for
statistical analysis using <em>R/Bioconductor</em> packages, such as <code>DESeq2</code>
or <code>edgeR</code>.</p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="bulk-rnaseq.html#cb196-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(samples)) {</span>
<span id="cb196-2"><a href="bulk-rnaseq.html#cb196-2"></a>    exp1 &lt;-<span class="st"> </span><span class="kw">rExp</span>(countfiles[i])</span>
<span id="cb196-3"><a href="bulk-rnaseq.html#cb196-3"></a>    <span class="kw">write.table</span>(exp1, <span class="dt">file =</span> <span class="kw">paste0</span>(<span class="st">&quot;output/RNAseq_bulk&quot;</span>, samples[i],</span>
<span id="cb196-4"><a href="bulk-rnaseq.html#cb196-4"></a>                                    <span class="st">&quot;/&quot;</span>, samples[i], <span class="st">&quot;_abundance.tsv&quot;</span>),</span>
<span id="cb196-5"><a href="bulk-rnaseq.html#cb196-5"></a>                <span class="dt">row.names =</span> <span class="ot">FALSE</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</span>
<span id="cb196-6"><a href="bulk-rnaseq.html#cb196-6"></a>}</span></code></pre></div>
</div>
<div id="transcriptome-quantification" class="section level2" number="9.5">
<h2><span class="header-section-number">9.5</span> transcriptome quantification</h2>
<p>There are many tools available for transcriptome quantification, such
as kallisto, StringTie, salmon, and Trinity. Here show the usage of
<code>Kallisto</code> and <code>salmon</code>.</p>
<div id="kallisto" class="section level3" number="9.5.1">
<h3><span class="header-section-number">9.5.1</span> Kallisto</h3>
<p>The kallisto is a tool to quantify transcript abundance with raw fastq
reads and indexed reference transcriptomes. The <code>Rcwl</code> tool of
<code>kallisto_index</code> below is used to build an index file from reference
transcriptome fasta.</p>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="bulk-rnaseq.html#cb197-1"></a>kallisto_index &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="st">&quot;tl_kallisto_index&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="bulk-rnaseq.html#cb198-1"></a><span class="kw">inputs</span>(kallisto_index)</span>
<span id="cb198-2"><a href="bulk-rnaseq.html#cb198-2"></a>kallisto_index<span class="op">$</span>fasta &lt;-<span class="st"> &quot;data/resources/gencode.v33.transcripts.fa&quot;</span></span>
<span id="cb198-3"><a href="bulk-rnaseq.html#cb198-3"></a>kallisto_index<span class="op">$</span>index &lt;-<span class="st"> &quot;gencode.v33.transcripts_kallisto&quot;</span></span>
<span id="cb198-4"><a href="bulk-rnaseq.html#cb198-4"></a><span class="kw">runCWL</span>(kallisto_index, <span class="dt">outdir =</span> <span class="st">&quot;data/resources/kallisto&quot;</span>, <span class="dt">showLog =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>The <code>Rcwl</code> tool <code>kallisto_quant</code> runs the quantification algorithm to
estimate the transcripts expression abundances.</p>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="bulk-rnaseq.html#cb199-1"></a>kallisto_quant &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="st">&quot;tl_kallisto_quant&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="bulk-rnaseq.html#cb200-1"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">fastq =</span> files)</span>
<span id="cb200-2"><a href="bulk-rnaseq.html#cb200-2"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">index =</span> <span class="st">&quot;data/resources/gencode.v33.transcripts_kallisto&quot;</span>,</span>
<span id="cb200-3"><a href="bulk-rnaseq.html#cb200-3"></a>                  <span class="dt">threads =</span> <span class="dv">16</span>)</span>
<span id="cb200-4"><a href="bulk-rnaseq.html#cb200-4"></a>bpparam &lt;-<span class="st"> </span><span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="kw">length</span>(samples),</span>
<span id="cb200-5"><a href="bulk-rnaseq.html#cb200-5"></a>                           <span class="dt">cluster =</span> <span class="st">&quot;multicore&quot;</span>)</span>
<span id="cb200-6"><a href="bulk-rnaseq.html#cb200-6"></a>                           </span>
<span id="cb200-7"><a href="bulk-rnaseq.html#cb200-7"></a>res2 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(kallisto_quant, <span class="dt">outdir =</span> <span class="st">&quot;output/RNAseq_bulk/kallisto&quot;</span>,</span>
<span id="cb200-8"><a href="bulk-rnaseq.html#cb200-8"></a>                    inputList, paramList,</span>
<span id="cb200-9"><a href="bulk-rnaseq.html#cb200-9"></a>                    <span class="dt">BPPARAM =</span> bpparam,</span>
<span id="cb200-10"><a href="bulk-rnaseq.html#cb200-10"></a>                    <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>)</span></code></pre></div>
<p>Then the tool results are collected here:</p>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb201-1"><a href="bulk-rnaseq.html#cb201-1"></a><span class="kw">list.files</span>(<span class="st">&quot;output/RNAseq_bulk/kalisto&quot;</span>, <span class="st">&quot;abundance&quot;</span>)</span></code></pre></div>
<pre><code>## character(0)</code></pre>
</div>
<div id="salmon" class="section level3" number="9.5.2">
<h3><span class="header-section-number">9.5.2</span> salmon</h3>
<p>To be added.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="dnaseq-variant-calling.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="single-cell-rnaseq.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/09-RNAseq_bulk.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Rcwl-Pipelines.pdf", "Rcwl-Pipelines.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
