<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Bioinformatics tools and pipelines using R and CWL</title>
  <meta name="description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Bioinformatics tools and pipelines using R and CWL" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Bioinformatics tools and pipelines using R and CWL" />
  
  <meta name="twitter:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  

<meta name="author" content="Qiang Hu" />
<meta name="author" content="Qian Liu" />


<meta name="date" content="2021-02-27" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="intro.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/viz-1.8.2/viz.js"></script>
<link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="libs/grViz-binding-1.0.6.1/grViz.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Rcwl and Case Study</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#r-package-installation"><i class="fa fa-check"></i><b>0.1</b> R package installation</a></li>
<li class="chapter" data-level="0.2" data-path="index.html"><a href="index.html#system-requirements"><i class="fa fa-check"></i><b>0.2</b> System requirements</a></li>
<li class="chapter" data-level="0.3" data-path="index.html"><a href="index.html#docker"><i class="fa fa-check"></i><b>0.3</b> Docker</a></li>
<li class="chapter" data-level="0.4" data-path="index.html"><a href="index.html#structure-of-the-book"><i class="fa fa-check"></i><b>0.4</b> Structure of the book</a></li>
<li class="chapter" data-level="0.5" data-path="index.html"><a href="index.html#r-session-information"><i class="fa fa-check"></i><b>0.5</b> R session information</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="get-started.html"><a href="get-started.html"><i class="fa fa-check"></i><b>2</b> Get started</a></li>
<li class="chapter" data-level="3" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html"><i class="fa fa-check"></i><b>3</b> Wrap command line tools</a>
<ul>
<li class="chapter" data-level="3.1" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#input-parameters"><i class="fa fa-check"></i><b>3.1</b> Input Parameters</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#essential-input-parameters"><i class="fa fa-check"></i><b>3.1.1</b> Essential Input parameters</a></li>
<li class="chapter" data-level="3.1.2" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#array-inputs"><i class="fa fa-check"></i><b>3.1.2</b> Array Inputs</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#output-parameters"><i class="fa fa-check"></i><b>3.2</b> Output Parameters</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#capturing-output"><i class="fa fa-check"></i><b>3.2.1</b> Capturing Output</a></li>
<li class="chapter" data-level="3.2.2" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#array-outputs"><i class="fa fa-check"></i><b>3.2.2</b> Array Outputs</a></li>
<li class="chapter" data-level="3.2.3" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#standard-output"><i class="fa fa-check"></i><b>3.2.3</b> Standard output</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="writing-pipeline.html"><a href="writing-pipeline.html"><i class="fa fa-check"></i><b>4</b> Writing Pipeline</a>
<ul>
<li class="chapter" data-level="4.1" data-path="writing-pipeline.html"><a href="writing-pipeline.html#scattering-pipeline"><i class="fa fa-check"></i><b>4.1</b> Scattering pipeline</a></li>
<li class="chapter" data-level="4.2" data-path="writing-pipeline.html"><a href="writing-pipeline.html#pipeline-plot"><i class="fa fa-check"></i><b>4.2</b> Pipeline plot</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html"><i class="fa fa-check"></i><b>5</b> Tool/pipeline execution</a>
<ul>
<li class="chapter" data-level="5.1" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#running-tools-in-docker"><i class="fa fa-check"></i><b>5.1</b> Running Tools in Docker</a></li>
<li class="chapter" data-level="5.2" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#running-tools-in-cluster-server"><i class="fa fa-check"></i><b>5.2</b> Running Tools in Cluster server</a></li>
<li class="chapter" data-level="5.3" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#web-application"><i class="fa fa-check"></i><b>5.3</b> Web Application</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#cwlprocess-example"><i class="fa fa-check"></i><b>5.3.1</b> cwlProcess example</a></li>
<li class="chapter" data-level="5.3.2" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#cwlprocess-to-shiny-app"><i class="fa fa-check"></i><b>5.3.2</b> cwlProcess to Shiny App</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html"><i class="fa fa-check"></i><b>6</b> RcwlPipelines</a>
<ul>
<li class="chapter" data-level="6.1" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#rcwl-recipes-and-cwl-scripts"><i class="fa fa-check"></i><b>6.1</b> Rcwl recipes and CWL scripts</a></li>
<li class="chapter" data-level="6.2" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#rcwlpipelines-core-functions"><i class="fa fa-check"></i><b>6.2</b> RcwlPipelines core functions</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#cwlupdate"><i class="fa fa-check"></i><b>6.2.1</b> <code>cwlUpdate</code></a></li>
<li class="chapter" data-level="6.2.2" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#cwlsearch"><i class="fa fa-check"></i><b>6.2.2</b> <code>cwlSearch</code></a></li>
<li class="chapter" data-level="6.2.3" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#cwlload"><i class="fa fa-check"></i><b>6.2.3</b> <code>cwlLoad</code></a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#toolpipeline-customization"><i class="fa fa-check"></i><b>6.3</b> Tool/pipeline customization</a></li>
<li class="chapter" data-level="6.4" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#build-a-pipeline"><i class="fa fa-check"></i><b>6.4</b> Build a pipeline</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="dnaseq-alignment.html"><a href="dnaseq-alignment.html"><i class="fa fa-check"></i><b>7</b> DNAseq alignment</a></li>
<li class="chapter" data-level="8" data-path="dnaseq-variant-calling.html"><a href="dnaseq-variant-calling.html"><i class="fa fa-check"></i><b>8</b> DNAseq variant calling</a></li>
<li class="chapter" data-level="9" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html"><i class="fa fa-check"></i><b>9</b> Bulk RNAseq</a>
<ul>
<li class="chapter" data-level="9.1" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#prepare-data"><i class="fa fa-check"></i><b>9.1</b> Prepare data</a></li>
<li class="chapter" data-level="9.2" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#submit-parallel-jobs"><i class="fa fa-check"></i><b>9.2</b> Submit parallel jobs</a></li>
<li class="chapter" data-level="9.3" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#qc-summary"><i class="fa fa-check"></i><b>9.3</b> QC Summary</a></li>
<li class="chapter" data-level="9.4" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#abundances-summary"><i class="fa fa-check"></i><b>9.4</b> Abundances summary</a></li>
<li class="chapter" data-level="9.5" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#transcriptome-quantification"><i class="fa fa-check"></i><b>9.5</b> transcriptome quantification</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#kallisto"><i class="fa fa-check"></i><b>9.5.1</b> Kallisto</a></li>
<li class="chapter" data-level="9.5.2" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#salmon"><i class="fa fa-check"></i><b>9.5.2</b> salmon</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="single-cell-rnaseq.html"><a href="single-cell-rnaseq.html"><i class="fa fa-check"></i><b>10</b> single cell RNAseq</a></li>
<li class="chapter" data-level="11" data-path="mirna.html"><a href="mirna.html"><i class="fa fa-check"></i><b>11</b> miRNA</a>
<ul>
<li class="chapter" data-level="11.0.1" data-path="mirna.html"><a href="mirna.html#reference-index"><i class="fa fa-check"></i><b>11.0.1</b> Reference index</a></li>
<li class="chapter" data-level="11.0.2" data-path="mirna.html"><a href="mirna.html#run-mirdeep2-pipeline"><i class="fa fa-check"></i><b>11.0.2</b> Run miRDeep2 pipeline</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics tools and pipelines using R and CWL</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Bioinformatics tools and pipelines using R and CWL</h1>
<p class="author"><em>Qiang Hu</em></p>
<p class="author"><em>Qian Liu</em></p>
<p class="date"><em>2021-02-27</em></p>
</div>
<div id="preface" class="section level1 unnumbered" number="">
<h1>Preface</h1>
<p>This book introduces the <em>R/Bioconductor</em> packages,
<a href="https://bioconductor.org/packages/Rcwl/">Rcwl</a> and
<a href="https://bioconductor.org/packages/RcwlPipelines/">RcwlPipelines</a>, to
improve the way of building, managing and running Bioinformatics tools
and pipelines within <em>R</em>.</p>
<p>The <code>Rcwl</code> package is built on top of the Common Workflow Language
(CWL), and provides a simple and user-friendly way to wrap command
line tools into data analysis pipelines in <em>R</em>. The <code>RcwlPipelines</code>
package manages a collection of bioinformatics tools and pipelines
based on <code>Rcwl</code>.</p>
<div id="r-package-installation" class="section level2" number="0.1">
<h2><span class="header-section-number">0.1</span> R package installation</h2>
<p>The <code>Rcwl</code> and <code>RcwlPipelines</code> packages can be installed from
Bioconductor or Github:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="index.html#cb1-1"></a>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="kw">c</span>(<span class="st">&quot;Rcwl&quot;</span>, <span class="st">&quot;RcwlPipelines&quot;</span>))</span>
<span id="cb1-2"><a href="index.html#cb1-2"></a><span class="co"># or the development version</span></span>
<span id="cb1-3"><a href="index.html#cb1-3"></a>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="kw">c</span>(<span class="st">&quot;rworkflow/Rcwl&quot;</span>, <span class="st">&quot;rworkflow/RcwlPipelines&quot;</span>))</span></code></pre></div>
<p>To load the packages into R session:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="index.html#cb2-1"></a><span class="kw">library</span>(Rcwl)</span>
<span id="cb2-2"><a href="index.html#cb2-2"></a><span class="kw">library</span>(RcwlPipelines)</span></code></pre></div>
</div>
<div id="system-requirements" class="section level2" number="0.2">
<h2><span class="header-section-number">0.2</span> System requirements</h2>
<p>In addition to the <em>R</em> packages, the following tools are also required
to successfully run the tools/pipelines. If not locally available,
these tools will be installed automatically, powered by the <code>basilisk</code>
package.</p>
<ul>
<li>python (&gt;= 2.7)</li>
<li>cwltool (&gt;= 1.0.2018)</li>
<li>nodejs</li>
</ul>
<p>The <strong>cwltool</strong> is the reference implementation of the Common Workflow
Language, which is used to run the CWL scripts. The <strong>nodejs</strong> is
required when the CWL scripts use JavaScript. More details about these
tools can be found here:
* <a href="https://github.com/common-workflow-language/cwltool" class="uri">https://github.com/common-workflow-language/cwltool</a>
* <a href="https://nodejs.org" class="uri">https://nodejs.org</a></p>
</div>
<div id="docker" class="section level2" number="0.3">
<h2><span class="header-section-number">0.3</span> Docker</h2>
<p>The Docker container simplifies software installation and management,
especially for bioinformatics tools/pipelines requiring different
runtime environments and library dependencies. A CWL runner can
perform this work automatically by pulling the Docker containers and
mounting the paths of input files.</p>
<p>The Docker requirement is optional, as CWL scripts can also be run
locally with all the dependencies pre-installed.</p>
</div>
<div id="structure-of-the-book" class="section level2" number="0.4">
<h2><span class="header-section-number">0.4</span> Structure of the book</h2>
<ul>
<li>Introduction</li>
<li>Get started</li>
<li>Wrap command line tools</li>
<li>Writing Pipeline</li>
<li>Tool/pipeline execution</li>
<li>RcwlPipelines</li>
<li>DNAseq alignment</li>
<li>DNAseq variant calling</li>
<li>Bulk RNAseq</li>
<li>Single cell RNAseq</li>
<li>miRNA</li>
</ul>
</div>
<div id="r-session-information" class="section level2" number="0.5">
<h2><span class="header-section-number">0.5</span> R session information</h2>
<p>The R session information for compiling this mannual is shown below:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="index.html#cb3-1"></a><span class="kw">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: macOS Catalina 10.15.7
## 
## Matrix products: default
## BLAS/LAPACK: /Users/qi31566/miniconda3/envs/r-base/lib/libopenblasp-r0.3.12.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] bookdown_0.21               DropletUtils_1.10.3        
##  [3] SingleCellExperiment_1.12.0 SummarizedExperiment_1.20.0
##  [5] Biobase_2.50.0              GenomicRanges_1.42.0       
##  [7] GenomeInfoDb_1.26.2         IRanges_2.24.1             
##  [9] MatrixGenerics_1.2.1        matrixStats_0.58.0         
## [11] BiocStyle_2.18.1            BiocParallel_1.24.1        
## [13] RcwlPipelines_1.7.7         BiocFileCache_1.14.0       
## [15] dbplyr_2.1.0                Rcwl_1.7.12                
## [17] S4Vectors_0.28.1            BiocGenerics_0.36.0        
## [19] yaml_2.2.1                 
## 
## loaded via a namespace (and not attached):
##   [1] ellipsis_0.3.1            rprojroot_2.0.2          
##   [3] scuttle_1.0.4             XVector_0.30.0           
##   [5] fs_1.5.0                  rstudioapi_0.13          
##   [7] remotes_2.2.0             bit64_4.0.5              
##   [9] fansi_0.4.2               sparseMatrixStats_1.2.1  
##  [11] codetools_0.2-18          R.methodsS3_1.8.1        
##  [13] cachem_1.0.4              knitr_1.31               
##  [15] pkgload_1.2.0             jsonlite_1.7.2           
##  [17] R.oo_1.24.0               HDF5Array_1.18.1         
##  [19] shiny_1.6.0               DiagrammeR_1.0.6.1       
##  [21] BiocManager_1.30.10       compiler_4.0.3           
##  [23] httr_1.4.2                dqrng_0.2.1              
##  [25] basilisk_1.2.1            backports_1.2.1          
##  [27] assertthat_0.2.1          Matrix_1.3-2             
##  [29] fastmap_1.1.0             limma_3.46.0             
##  [31] cli_2.3.1                 later_1.1.0.1            
##  [33] visNetwork_2.0.9          htmltools_0.5.1.1        
##  [35] prettyunits_1.1.1         tools_4.0.3              
##  [37] igraph_1.2.6              glue_1.4.2               
##  [39] GenomeInfoDbData_1.2.4    dplyr_1.0.4              
##  [41] batchtools_0.9.15         rappdirs_0.3.3           
##  [43] tinytex_0.29              Rcpp_1.0.6               
##  [45] jquerylib_0.1.3           rhdf5filters_1.2.0       
##  [47] vctrs_0.3.6               DelayedMatrixStats_1.12.3
##  [49] xfun_0.21                 stringr_1.4.0            
##  [51] ps_1.5.0                  beachmat_2.6.4           
##  [53] testthat_3.0.2            mime_0.10                
##  [55] lifecycle_1.0.0           devtools_2.3.2           
##  [57] edgeR_3.32.1              zlibbioc_1.36.0          
##  [59] basilisk.utils_1.2.2      hms_1.0.0                
##  [61] promises_1.2.0.1          rhdf5_2.34.0             
##  [63] RColorBrewer_1.1-2        curl_4.3                 
##  [65] memoise_2.0.0             reticulate_1.18          
##  [67] sass_0.3.1                stringi_1.5.3            
##  [69] RSQLite_2.2.3             desc_1.2.0               
##  [71] checkmate_2.0.0           filelock_1.0.2           
##  [73] pkgbuild_1.2.0            rlang_0.4.10             
##  [75] pkgconfig_2.0.3           bitops_1.0-6             
##  [77] evaluate_0.14             lattice_0.20-41          
##  [79] Rhdf5lib_1.12.1           purrr_0.3.4              
##  [81] htmlwidgets_1.5.3         bit_4.0.4                
##  [83] processx_3.4.5            tidyselect_1.1.0         
##  [85] magrittr_2.0.1            R6_2.5.0                 
##  [87] generics_0.1.0            base64url_1.4            
##  [89] DelayedArray_0.16.1       DBI_1.1.1                
##  [91] pillar_1.5.0              withr_2.4.1              
##  [93] RCurl_1.98-1.2            tibble_3.0.6             
##  [95] crayon_1.4.1              utf8_1.1.4               
##  [97] rmarkdown_2.7             progress_1.2.2           
##  [99] usethis_2.0.1             locfit_1.5-9.4           
## [101] grid_4.0.3                data.table_1.14.0        
## [103] blob_1.2.1                callr_3.5.1              
## [105] git2r_0.28.0              digest_0.6.27            
## [107] xtable_1.8-4              tidyr_1.1.2              
## [109] httpuv_1.5.5              brew_1.0-6               
## [111] R.utils_2.10.1            bslib_0.2.4              
## [113] sessioninfo_1.1.1</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="intro.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/index.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Rcwl-Pipelines.pdf", "Rcwl-Pipelines.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
