<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 11 miRNA | Bioinformatics tools and pipelines using R and CWL</title>
  <meta name="description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 11 miRNA | Bioinformatics tools and pipelines using R and CWL" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 11 miRNA | Bioinformatics tools and pipelines using R and CWL" />
  
  <meta name="twitter:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  

<meta name="author" content="Qiang Hu" />
<meta name="author" content="Qian Liu" />


<meta name="date" content="2021-02-27" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="single-cell-rnaseq.html"/>
<link rel="next" href="references.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/viz-1.8.2/viz.js"></script>
<link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="libs/grViz-binding-1.0.6.1/grViz.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Rcwl and Case Study</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#r-package-installation"><i class="fa fa-check"></i><b>0.1</b> R package installation</a></li>
<li class="chapter" data-level="0.2" data-path="index.html"><a href="index.html#system-requirements"><i class="fa fa-check"></i><b>0.2</b> System requirements</a></li>
<li class="chapter" data-level="0.3" data-path="index.html"><a href="index.html#docker"><i class="fa fa-check"></i><b>0.3</b> Docker</a></li>
<li class="chapter" data-level="0.4" data-path="index.html"><a href="index.html#structure-of-the-book"><i class="fa fa-check"></i><b>0.4</b> Structure of the book</a></li>
<li class="chapter" data-level="0.5" data-path="index.html"><a href="index.html#r-session-information"><i class="fa fa-check"></i><b>0.5</b> R session information</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="get-started.html"><a href="get-started.html"><i class="fa fa-check"></i><b>2</b> Get started</a></li>
<li class="chapter" data-level="3" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html"><i class="fa fa-check"></i><b>3</b> Wrap command line tools</a>
<ul>
<li class="chapter" data-level="3.1" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#input-parameters"><i class="fa fa-check"></i><b>3.1</b> Input Parameters</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#essential-input-parameters"><i class="fa fa-check"></i><b>3.1.1</b> Essential Input parameters</a></li>
<li class="chapter" data-level="3.1.2" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#array-inputs"><i class="fa fa-check"></i><b>3.1.2</b> Array Inputs</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#output-parameters"><i class="fa fa-check"></i><b>3.2</b> Output Parameters</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#capturing-output"><i class="fa fa-check"></i><b>3.2.1</b> Capturing Output</a></li>
<li class="chapter" data-level="3.2.2" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#array-outputs"><i class="fa fa-check"></i><b>3.2.2</b> Array Outputs</a></li>
<li class="chapter" data-level="3.2.3" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#standard-output"><i class="fa fa-check"></i><b>3.2.3</b> Standard output</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="writing-pipeline.html"><a href="writing-pipeline.html"><i class="fa fa-check"></i><b>4</b> Writing Pipeline</a>
<ul>
<li class="chapter" data-level="4.1" data-path="writing-pipeline.html"><a href="writing-pipeline.html#scattering-pipeline"><i class="fa fa-check"></i><b>4.1</b> Scattering pipeline</a></li>
<li class="chapter" data-level="4.2" data-path="writing-pipeline.html"><a href="writing-pipeline.html#pipeline-plot"><i class="fa fa-check"></i><b>4.2</b> Pipeline plot</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html"><i class="fa fa-check"></i><b>5</b> Tool/pipeline execution</a>
<ul>
<li class="chapter" data-level="5.1" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#running-tools-in-docker"><i class="fa fa-check"></i><b>5.1</b> Running Tools in Docker</a></li>
<li class="chapter" data-level="5.2" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#running-tools-in-cluster-server"><i class="fa fa-check"></i><b>5.2</b> Running Tools in Cluster server</a></li>
<li class="chapter" data-level="5.3" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#web-application"><i class="fa fa-check"></i><b>5.3</b> Web Application</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#cwlprocess-example"><i class="fa fa-check"></i><b>5.3.1</b> cwlProcess example</a></li>
<li class="chapter" data-level="5.3.2" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#cwlprocess-to-shiny-app"><i class="fa fa-check"></i><b>5.3.2</b> cwlProcess to Shiny App</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html"><i class="fa fa-check"></i><b>6</b> RcwlPipelines</a>
<ul>
<li class="chapter" data-level="6.1" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#rcwl-recipes-and-cwl-scripts"><i class="fa fa-check"></i><b>6.1</b> Rcwl recipes and CWL scripts</a></li>
<li class="chapter" data-level="6.2" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#rcwlpipelines-core-functions"><i class="fa fa-check"></i><b>6.2</b> RcwlPipelines core functions</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#cwlupdate"><i class="fa fa-check"></i><b>6.2.1</b> <code>cwlUpdate</code></a></li>
<li class="chapter" data-level="6.2.2" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#cwlsearch"><i class="fa fa-check"></i><b>6.2.2</b> <code>cwlSearch</code></a></li>
<li class="chapter" data-level="6.2.3" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#cwlload"><i class="fa fa-check"></i><b>6.2.3</b> <code>cwlLoad</code></a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#toolpipeline-customization"><i class="fa fa-check"></i><b>6.3</b> Tool/pipeline customization</a></li>
<li class="chapter" data-level="6.4" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#build-a-pipeline"><i class="fa fa-check"></i><b>6.4</b> Build a pipeline</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="dnaseq-alignment.html"><a href="dnaseq-alignment.html"><i class="fa fa-check"></i><b>7</b> DNAseq alignment</a></li>
<li class="chapter" data-level="8" data-path="dnaseq-variant-calling.html"><a href="dnaseq-variant-calling.html"><i class="fa fa-check"></i><b>8</b> DNAseq variant calling</a></li>
<li class="chapter" data-level="9" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html"><i class="fa fa-check"></i><b>9</b> Bulk RNAseq</a>
<ul>
<li class="chapter" data-level="9.1" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#prepare-data"><i class="fa fa-check"></i><b>9.1</b> Prepare data</a></li>
<li class="chapter" data-level="9.2" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#submit-parallel-jobs"><i class="fa fa-check"></i><b>9.2</b> Submit parallel jobs</a></li>
<li class="chapter" data-level="9.3" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#qc-summary"><i class="fa fa-check"></i><b>9.3</b> QC Summary</a></li>
<li class="chapter" data-level="9.4" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#abundances-summary"><i class="fa fa-check"></i><b>9.4</b> Abundances summary</a></li>
<li class="chapter" data-level="9.5" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#transcriptome-quantification"><i class="fa fa-check"></i><b>9.5</b> transcriptome quantification</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#kallisto"><i class="fa fa-check"></i><b>9.5.1</b> Kallisto</a></li>
<li class="chapter" data-level="9.5.2" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#salmon"><i class="fa fa-check"></i><b>9.5.2</b> salmon</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="single-cell-rnaseq.html"><a href="single-cell-rnaseq.html"><i class="fa fa-check"></i><b>10</b> single cell RNAseq</a></li>
<li class="chapter" data-level="11" data-path="mirna.html"><a href="mirna.html"><i class="fa fa-check"></i><b>11</b> miRNA</a>
<ul>
<li class="chapter" data-level="11.0.1" data-path="mirna.html"><a href="mirna.html#reference-index"><i class="fa fa-check"></i><b>11.0.1</b> Reference index</a></li>
<li class="chapter" data-level="11.0.2" data-path="mirna.html"><a href="mirna.html#run-mirdeep2-pipeline"><i class="fa fa-check"></i><b>11.0.2</b> Run miRDeep2 pipeline</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics tools and pipelines using R and CWL</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="mirna" class="section level1" number="11">
<h1><span class="header-section-number">Chapter 11</span> miRNA</h1>
<p>The <code>miRDeep2</code> is one of the most popular tools for discovering known
and novel miRNAs from small RNA sequencing data. We have wrapped the
mapping and quantification steps into an <code>Rcwl</code> pipeline which is
ready to load and use. More details about <code>miRDeep2</code> can be found
here: <a href="https://github.com/rajewsky-lab/mirdeep2" class="uri">https://github.com/rajewsky-lab/mirdeep2</a>.</p>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="mirna.html#cb203-1"></a>miRDeep2PL &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="st">&quot;pl_miRDeep2PL&quot;</span>)</span></code></pre></div>
<pre><code>## miRMapper loaded</code></pre>
<pre><code>## miRDeep2 loaded</code></pre>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="mirna.html#cb206-1"></a><span class="kw">plotCWL</span>(miRDeep2PL)</span></code></pre></div>
<div id="htmlwidget-016e725c776b30c5a75f" style="width:672px;height:480px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-016e725c776b30c5a75f">{"x":{"diagram":"digraph {\n\ngraph [layout = \"neato\",\n       outputorder = \"edgesfirst\",\n       bgcolor = \"white\"]\n\nnode [fontname = \"Helvetica\",\n      fontsize = \"10\",\n      shape = \"circle\",\n      fixedsize = \"true\",\n      width = \"0.5\",\n      style = \"filled\",\n      fillcolor = \"aliceblue\",\n      color = \"gray70\",\n      fontcolor = \"gray50\"]\n\nedge [fontname = \"Helvetica\",\n     fontsize = \"8\",\n     len = \"1.5\",\n     color = \"gray80\",\n     arrowsize = \"0.5\"]\n\n  \"1\" [label = \"reads\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"3.5,5!\"] \n  \"2\" [label = \"format\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,5!\"] \n  \"3\" [label = \"adapter\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2.5,5!\"] \n  \"4\" [label = \"len\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"8.5,5!\"] \n  \"5\" [label = \"genome\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1,5!\"] \n  \"6\" [label = \"miRef\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"4.5,5!\"] \n  \"7\" [label = \"miOther\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"5.5,5!\"] \n  \"8\" [label = \"precursors\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"6.5,5!\"] \n  \"9\" [label = \"species\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"7.5,5!\"] \n  \"10\" [label = \"csvfiles\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1.5,1!\"] \n  \"11\" [label = \"htmls\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2.5,1!\"] \n  \"12\" [label = \"bed\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"3.5,1!\"] \n  \"13\" [label = \"expression\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"4.5,1!\"] \n  \"14\" [label = \"mirna_results\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"5.5,1!\"] \n  \"15\" [label = \"pdfs\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"6.5,1!\"] \n  \"16\" [label = \"preads\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1.5,3!\"] \n  \"17\" [label = \"arf\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2.5,3!\"] \n  \"18\" [label = \"Mapper\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1.5,4!\"] \n  \"19\" [label = \"miRDeep2\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"3.5,2!\"] \n  \"1\"->\"18\" \n  \"2\"->\"18\" \n  \"3\"->\"18\" \n  \"5\"->\"18\" \n  \"18\"->\"16\" \n  \"18\"->\"17\" \n  \"16\"->\"19\" \n  \"5\"->\"19\" \n  \"17\"->\"19\" \n  \"6\"->\"19\" \n  \"7\"->\"19\" \n  \"8\"->\"19\" \n  \"9\"->\"19\" \n  \"19\"->\"10\" \n  \"19\"->\"11\" \n  \"19\"->\"12\" \n  \"19\"->\"13\" \n  \"19\"->\"14\" \n  \"19\"->\"15\" \n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>
<p>Here We also use the data from the above GitHub repository as an example.
<a href="https://github.com/rajewsky-lab/mirdeep2/tree/master/tutorial_dir" class="uri">https://github.com/rajewsky-lab/mirdeep2/tree/master/tutorial_dir</a></p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="mirna.html#cb207-1"></a>git2r<span class="op">::</span><span class="kw">clone</span>(<span class="st">&quot;https://github.com/rajewsky-lab/mirdeep2&quot;</span>, <span class="st">&quot;data/miRNA&quot;</span>)</span>
<span id="cb207-2"><a href="mirna.html#cb207-2"></a><span class="kw">list.files</span>(<span class="st">&quot;data/miRNA/tutorial_dir&quot;</span>)</span></code></pre></div>
<div id="reference-index" class="section level3" number="11.0.1">
<h3><span class="header-section-number">11.0.1</span> Reference index</h3>
<p>First, we need to build indexes for the miRNA reference with the
<code>Rcwl</code> tool <code>bowtie_build</code>. This is only required to be performed once
for each refernce genome.</p>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="mirna.html#cb208-1"></a>bowtie_build &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="st">&quot;tl_bowtie_build&quot;</span>)</span>
<span id="cb208-2"><a href="mirna.html#cb208-2"></a><span class="kw">inputs</span>(bowtie_build)</span></code></pre></div>
<pre><code>## inputs:
##   ref (File):  
##   outPrefix (string):</code></pre>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb210-1"><a href="mirna.html#cb210-1"></a>bowtie_build<span class="op">$</span>ref &lt;-<span class="st"> &quot;data/miRNA/tutorial_dir/cel_cluster.fa&quot;</span></span>
<span id="cb210-2"><a href="mirna.html#cb210-2"></a>bowtie_build<span class="op">$</span>outPrefix &lt;-<span class="st"> &quot;cel_cluster&quot;</span></span>
<span id="cb210-3"><a href="mirna.html#cb210-3"></a>idxRes &lt;-<span class="st"> </span><span class="kw">runCWL</span>(bowtie_build, <span class="dt">outdir =</span> <span class="st">&quot;output/miRNA/genome&quot;</span>,</span>
<span id="cb210-4"><a href="mirna.html#cb210-4"></a>                 <span class="dt">showLog =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;output/miRNA&quot;</span>)</span></code></pre></div>
<p>Then the indexed reference files are generated in the output directory
defined in <code>outdir</code>.</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="mirna.html#cb211-1"></a><span class="kw">file.copy</span>(<span class="st">&quot;data/miRNA/tutorial_dir/cel_cluster.fa&quot;</span>,</span>
<span id="cb211-2"><a href="mirna.html#cb211-2"></a>          <span class="st">&quot;output/miRNA/genome/cel_cluster.fa&quot;</span>)</span></code></pre></div>
<pre><code>## Warning in file.create(to[okay]): cannot create file &#39;output/miRNA/genome/
## cel_cluster.fa&#39;, reason &#39;No such file or directory&#39;</code></pre>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="mirna.html#cb214-1"></a><span class="kw">dir</span>(<span class="st">&quot;output/miRNA/genome&quot;</span>)</span></code></pre></div>
<pre><code>## character(0)</code></pre>
</div>
<div id="run-mirdeep2-pipeline" class="section level3" number="11.0.2">
<h3><span class="header-section-number">11.0.2</span> Run miRDeep2 pipeline</h3>
<p>To run the pipeline for all samples parallelly, we need to prepare the
inputs for arguments of <code>inputList</code> and <code>paramList</code>.</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="mirna.html#cb216-1"></a><span class="kw">inputs</span>(miRDeep2PL)</span></code></pre></div>
<pre><code>## inputs:
##   reads (File):  
##   format (string):  -c
##   adapter (string):  
##   len (int):  18
##   genome (File):  
##   miRef  ( File|string ): 
##   miOther  ( File|string ): 
##   precursors  ( File|string ): 
##   species (string):</code></pre>
<p>To mimic multiple samples, here we just repeat to use the input reads
as if they are two different samples.</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="mirna.html#cb218-1"></a>reads &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">sample1 =</span> <span class="st">&quot;data/miRNA/tutorial_dir/reads.fa&quot;</span>,</span>
<span id="cb218-2"><a href="mirna.html#cb218-2"></a>              <span class="dt">sample2 =</span> <span class="st">&quot;data/miRNA/tutorial_dir/reads.fa&quot;</span>)</span>
<span id="cb218-3"><a href="mirna.html#cb218-3"></a></span>
<span id="cb218-4"><a href="mirna.html#cb218-4"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">reads =</span> reads)</span>
<span id="cb218-5"><a href="mirna.html#cb218-5"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">adapter =</span> <span class="st">&quot;TCGTATGCCGTCTTCTGCTTGT&quot;</span>,</span>
<span id="cb218-6"><a href="mirna.html#cb218-6"></a>                  <span class="dt">genome =</span> <span class="st">&quot;output/miRNA/genome/cel_cluster.fa&quot;</span>,</span>
<span id="cb218-7"><a href="mirna.html#cb218-7"></a>                  <span class="dt">miRef =</span> <span class="st">&quot;data/miRNA/tutorial_dir/mature_ref_this_species.fa&quot;</span>,</span>
<span id="cb218-8"><a href="mirna.html#cb218-8"></a>                  <span class="dt">miOther =</span> <span class="st">&quot;data/miRNA/tutorial_dir/mature_ref_other_species.fa&quot;</span>,</span>
<span id="cb218-9"><a href="mirna.html#cb218-9"></a>                  <span class="dt">precursors =</span> <span class="st">&quot;data/miRNA/tutorial_dir/precursors_ref_this_species.fa&quot;</span>,</span>
<span id="cb218-10"><a href="mirna.html#cb218-10"></a>                  <span class="dt">species =</span> <span class="st">&quot;C.elegans&quot;</span>)</span></code></pre></div>
<p>Let’s run the pipeline with two computing nodes.</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="mirna.html#cb219-1"></a>mirRes &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(miRDeep2PL, <span class="dt">outdir =</span> <span class="st">&quot;output/miRNA&quot;</span>,</span>
<span id="cb219-2"><a href="mirna.html#cb219-2"></a>                      inputList, paramList,</span>
<span id="cb219-3"><a href="mirna.html#cb219-3"></a>                      <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(</span>
<span id="cb219-4"><a href="mirna.html#cb219-4"></a>                          <span class="dt">workers =</span> <span class="dv">2</span>, <span class="dt">cluster =</span> <span class="st">&quot;multicore&quot;</span>))</span></code></pre></div>
<p>The results are collected in the output directory defined in the <code>outdir</code>.</p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="mirna.html#cb220-1"></a><span class="kw">dir</span>(<span class="st">&quot;output/miRNA/sample1&quot;</span>)</span></code></pre></div>
<pre><code>## character(0)</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="single-cell-rnaseq.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/11-miRNA.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Rcwl-Pipelines.pdf", "Rcwl-Pipelines.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
