<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 RcwlPipelines | Bioinformatics tools and pipelines using R and CWL</title>
  <meta name="description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 RcwlPipelines | Bioinformatics tools and pipelines using R and CWL" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 RcwlPipelines | Bioinformatics tools and pipelines using R and CWL" />
  
  <meta name="twitter:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  

<meta name="author" content="Qiang Hu" />
<meta name="author" content="Qian Liu" />


<meta name="date" content="2021-02-27" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="toolpipeline-execution.html"/>
<link rel="next" href="dnaseq-alignment.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/viz-1.8.2/viz.js"></script>
<link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="libs/grViz-binding-1.0.6.1/grViz.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Rcwl and Case Study</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#r-package-installation"><i class="fa fa-check"></i><b>0.1</b> R package installation</a></li>
<li class="chapter" data-level="0.2" data-path="index.html"><a href="index.html#system-requirements"><i class="fa fa-check"></i><b>0.2</b> System requirements</a></li>
<li class="chapter" data-level="0.3" data-path="index.html"><a href="index.html#docker"><i class="fa fa-check"></i><b>0.3</b> Docker</a></li>
<li class="chapter" data-level="0.4" data-path="index.html"><a href="index.html#structure-of-the-book"><i class="fa fa-check"></i><b>0.4</b> Structure of the book</a></li>
<li class="chapter" data-level="0.5" data-path="index.html"><a href="index.html#r-session-information"><i class="fa fa-check"></i><b>0.5</b> R session information</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="get-started.html"><a href="get-started.html"><i class="fa fa-check"></i><b>2</b> Get started</a></li>
<li class="chapter" data-level="3" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html"><i class="fa fa-check"></i><b>3</b> Wrap command line tools</a>
<ul>
<li class="chapter" data-level="3.1" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#input-parameters"><i class="fa fa-check"></i><b>3.1</b> Input Parameters</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#essential-input-parameters"><i class="fa fa-check"></i><b>3.1.1</b> Essential Input parameters</a></li>
<li class="chapter" data-level="3.1.2" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#array-inputs"><i class="fa fa-check"></i><b>3.1.2</b> Array Inputs</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#output-parameters"><i class="fa fa-check"></i><b>3.2</b> Output Parameters</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#capturing-output"><i class="fa fa-check"></i><b>3.2.1</b> Capturing Output</a></li>
<li class="chapter" data-level="3.2.2" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#array-outputs"><i class="fa fa-check"></i><b>3.2.2</b> Array Outputs</a></li>
<li class="chapter" data-level="3.2.3" data-path="wrap-command-line-tools.html"><a href="wrap-command-line-tools.html#standard-output"><i class="fa fa-check"></i><b>3.2.3</b> Standard output</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="writing-pipeline.html"><a href="writing-pipeline.html"><i class="fa fa-check"></i><b>4</b> Writing Pipeline</a>
<ul>
<li class="chapter" data-level="4.1" data-path="writing-pipeline.html"><a href="writing-pipeline.html#scattering-pipeline"><i class="fa fa-check"></i><b>4.1</b> Scattering pipeline</a></li>
<li class="chapter" data-level="4.2" data-path="writing-pipeline.html"><a href="writing-pipeline.html#pipeline-plot"><i class="fa fa-check"></i><b>4.2</b> Pipeline plot</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html"><i class="fa fa-check"></i><b>5</b> Tool/pipeline execution</a>
<ul>
<li class="chapter" data-level="5.1" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#running-tools-in-docker"><i class="fa fa-check"></i><b>5.1</b> Running Tools in Docker</a></li>
<li class="chapter" data-level="5.2" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#running-tools-in-cluster-server"><i class="fa fa-check"></i><b>5.2</b> Running Tools in Cluster server</a></li>
<li class="chapter" data-level="5.3" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#web-application"><i class="fa fa-check"></i><b>5.3</b> Web Application</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#cwlprocess-example"><i class="fa fa-check"></i><b>5.3.1</b> cwlProcess example</a></li>
<li class="chapter" data-level="5.3.2" data-path="toolpipeline-execution.html"><a href="toolpipeline-execution.html#cwlprocess-to-shiny-app"><i class="fa fa-check"></i><b>5.3.2</b> cwlProcess to Shiny App</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html"><i class="fa fa-check"></i><b>6</b> RcwlPipelines</a>
<ul>
<li class="chapter" data-level="6.1" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#rcwl-recipes-and-cwl-scripts"><i class="fa fa-check"></i><b>6.1</b> Rcwl recipes and CWL scripts</a></li>
<li class="chapter" data-level="6.2" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#rcwlpipelines-core-functions"><i class="fa fa-check"></i><b>6.2</b> RcwlPipelines core functions</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#cwlupdate"><i class="fa fa-check"></i><b>6.2.1</b> <code>cwlUpdate</code></a></li>
<li class="chapter" data-level="6.2.2" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#cwlsearch"><i class="fa fa-check"></i><b>6.2.2</b> <code>cwlSearch</code></a></li>
<li class="chapter" data-level="6.2.3" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#cwlload"><i class="fa fa-check"></i><b>6.2.3</b> <code>cwlLoad</code></a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#toolpipeline-customization"><i class="fa fa-check"></i><b>6.3</b> Tool/pipeline customization</a></li>
<li class="chapter" data-level="6.4" data-path="rcwlpipelines.html"><a href="rcwlpipelines.html#build-a-pipeline"><i class="fa fa-check"></i><b>6.4</b> Build a pipeline</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="dnaseq-alignment.html"><a href="dnaseq-alignment.html"><i class="fa fa-check"></i><b>7</b> DNAseq alignment</a></li>
<li class="chapter" data-level="8" data-path="dnaseq-variant-calling.html"><a href="dnaseq-variant-calling.html"><i class="fa fa-check"></i><b>8</b> DNAseq variant calling</a></li>
<li class="chapter" data-level="9" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html"><i class="fa fa-check"></i><b>9</b> Bulk RNAseq</a>
<ul>
<li class="chapter" data-level="9.1" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#prepare-data"><i class="fa fa-check"></i><b>9.1</b> Prepare data</a></li>
<li class="chapter" data-level="9.2" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#submit-parallel-jobs"><i class="fa fa-check"></i><b>9.2</b> Submit parallel jobs</a></li>
<li class="chapter" data-level="9.3" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#qc-summary"><i class="fa fa-check"></i><b>9.3</b> QC Summary</a></li>
<li class="chapter" data-level="9.4" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#abundances-summary"><i class="fa fa-check"></i><b>9.4</b> Abundances summary</a></li>
<li class="chapter" data-level="9.5" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#transcriptome-quantification"><i class="fa fa-check"></i><b>9.5</b> transcriptome quantification</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#kallisto"><i class="fa fa-check"></i><b>9.5.1</b> Kallisto</a></li>
<li class="chapter" data-level="9.5.2" data-path="bulk-rnaseq.html"><a href="bulk-rnaseq.html#salmon"><i class="fa fa-check"></i><b>9.5.2</b> salmon</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="single-cell-rnaseq.html"><a href="single-cell-rnaseq.html"><i class="fa fa-check"></i><b>10</b> single cell RNAseq</a></li>
<li class="chapter" data-level="11" data-path="mirna.html"><a href="mirna.html"><i class="fa fa-check"></i><b>11</b> miRNA</a>
<ul>
<li class="chapter" data-level="11.0.1" data-path="mirna.html"><a href="mirna.html#reference-index"><i class="fa fa-check"></i><b>11.0.1</b> Reference index</a></li>
<li class="chapter" data-level="11.0.2" data-path="mirna.html"><a href="mirna.html#run-mirdeep2-pipeline"><i class="fa fa-check"></i><b>11.0.2</b> Run miRDeep2 pipeline</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics tools and pipelines using R and CWL</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="rcwlpipelines" class="section level1" number="6">
<h1><span class="header-section-number">Chapter 6</span> RcwlPipelines</h1>
<div id="rcwl-recipes-and-cwl-scripts" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Rcwl recipes and CWL scripts</h2>
<p>The <em>R</em> scripts to build the CWL tools and pipelines are now residing
in a dedicated <a href="https://github.com/rworkflow/RcwlRecipes">GitHub
repository</a>, which is
intended to be a community effort to collect and contribute
Bioinformatics tools and pipelines using <code>Rcwl</code> and CWL. Script names
are prefixed with <code>tl_</code> and <code>pl_</code> for tools and pipelines
respectively.</p>
</div>
<div id="rcwlpipelines-core-functions" class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> RcwlPipelines core functions</h2>
<div id="cwlupdate" class="section level3" number="6.2.1">
<h3><span class="header-section-number">6.2.1</span> <code>cwlUpdate</code></h3>
<p>The <code>cwlUpdate</code> function syncs the current <code>Rcwl</code> recipes and returns
a <code>cwlHub</code> object which contains the most updated <code>Rcwl</code> recipes. The
<code>mcols()</code> function returns all related information about each
available tool or pipeline. Currently, we have integrated 113 command
line tools and 26 pipelines.</p>
<p>The recipes will be locally cached, so users don’t need to call
<code>cwlUpdate</code> every time unless they want to use a tool/pipeline that is
newly added to <code>RcwlPipelines</code>.</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="rcwlpipelines.html#cb104-1"></a>atls &lt;-<span class="st"> </span><span class="kw">cwlUpdate</span>(<span class="dt">branch =</span> <span class="st">&quot;dev&quot;</span>) <span class="co">## sync the tools/pipelines.</span></span>
<span id="cb104-2"><a href="rcwlpipelines.html#cb104-2"></a>atls</span></code></pre></div>
<pre><code>## cwlHub with 139 records
## cache path:  ~/Library/Caches/Rcwl 
## # last modified date:  2021-02-22 
## # cwlSearch() to query scripts
## # cwlLoad(&#39;title&#39;) to load the script
## # additional mcols(): rid, rpath, Type, Container, mtime, ...
## 
##             title                      
##   BFC2772 | pl_alignMerge              
##   BFC2773 | pl_AnnPhaseVcf             
##   BFC2774 | pl_BaseRecal               
##   BFC2775 | pl_bwaAlign                
##   BFC2776 | pl_bwaMMRecal              
##   ...       ...                        
##   BFC2906 | tl_VarScan2                
##   BFC2907 | tl_vcf_expression_annotator
##   BFC2908 | tl_vcf_readcount_annotator 
##   BFC2909 | tl_vep                     
##   BFC2910 | tl_vt_decompose            
##           Command                                                            
##   BFC2772 bwaAlign+mergeBamDup                                               
##   BFC2773 VCFvep+dVCFcoverage+rVCFcoverage+VCFexpression+PhaseVcf            
##   BFC2774 BaseRecalibrator+ApplyBQSR+samtools_index+samtools_flagstat+samt...
##   BFC2775 bwa+sam2bam+sortBam+idxBam                                         
##   BFC2776 bwaAlign+mergeBamDup+BaseRecal                                     
##   ...     ...                                                                
##   BFC2906                                                                    
##   BFC2907 vcf-expression-annotator                                           
##   BFC2908 vcf-readcount-annotator                                            
##   BFC2909 vep                                                                
##   BFC2910 vt decompose</code></pre>
<p>Currently, we have integrated 113 command
line tools and 26 pipelines.</p>
<pre><code>table(mcols(atls)$Type)</code></pre>
<p>We can also get the commands and docker containers for specific tool or pipeline.</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="rcwlpipelines.html#cb107-1"></a><span class="kw">mcols</span>(atls)[, <span class="kw">c</span>(<span class="st">&quot;Command&quot;</span>, <span class="st">&quot;Container&quot;</span>)]</span></code></pre></div>
<pre><code>## DataFrame with 139 rows and 2 columns
##                    Command              Container
##                &lt;character&gt;            &lt;character&gt;
## 1     bwaAlign+mergeBamDup                     NA
## 2   VCFvep+dVCFcoverage+..                     NA
## 3   BaseRecalibrator+App..                     NA
## 4   bwa+sam2bam+sortBam+..                     NA
## 5   bwaAlign+mergeBamDup..                     NA
## ...                    ...                    ...
## 135                        serge2016/varscan:v0..
## 136 vcf-expression-annot.. griffithlab/vatools:..
## 137 vcf-readcount-annota.. griffithlab/vatools:..
## 138                    vep hubentu/ensembl-vep-..
## 139           vt decompose             hubentu/vt</code></pre>
</div>
<div id="cwlsearch" class="section level3" number="6.2.2">
<h3><span class="header-section-number">6.2.2</span> <code>cwlSearch</code></h3>
<p>We can use (multiple) keywords to search for specific tools/pipelines
of interest, which internally search the <code>mcols</code> of “rname”, “rpath”,
“fpath”, “Command” and “Containers”. Here we show how to search the
alignment tool <code>bwa mem</code>.</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="rcwlpipelines.html#cb109-1"></a>t1 &lt;-<span class="st"> </span><span class="kw">cwlSearch</span>(<span class="kw">c</span>(<span class="st">&quot;bwa&quot;</span>, <span class="st">&quot;mem&quot;</span>))</span>
<span id="cb109-2"><a href="rcwlpipelines.html#cb109-2"></a>t1</span></code></pre></div>
<pre><code>## cwlHub with 1 records
## cache path:  ~/Library/Caches/Rcwl 
## # last modified date:  2021-02-22 
## # cwlSearch() to query scripts
## # cwlLoad(&#39;title&#39;) to load the script
## # additional mcols(): rid, rpath, Type, Container, mtime, ...
## 
##             title  Command
##   BFC2815 | tl_bwa bwa mem</code></pre>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="rcwlpipelines.html#cb111-1"></a><span class="kw">mcols</span>(t1)</span></code></pre></div>
<pre><code>## DataFrame with 1 row and 14 columns
##           rid       rname         create_time         access_time
##   &lt;character&gt; &lt;character&gt;         &lt;character&gt;         &lt;character&gt;
## 1     BFC2815      tl_bwa 2021-02-26 20:56:55 2021-02-26 20:56:55
##                    rpath       rtype                  fpath last_modified_time
##              &lt;character&gt; &lt;character&gt;            &lt;character&gt;          &lt;numeric&gt;
## 1 /Users/qi31566/Libra..       local /Users/qi31566/Libra..                 NA
##          etag   expires        Type     Command              Container
##   &lt;character&gt; &lt;numeric&gt; &lt;character&gt; &lt;character&gt;            &lt;character&gt;
## 1          NA        NA        tool     bwa mem biocontainers/bwa:v0..
##                 mtime
##           &lt;character&gt;
## 1 2021-02-22 14:45:16</code></pre>
</div>
<div id="cwlload" class="section level3" number="6.2.3">
<h3><span class="header-section-number">6.2.3</span> <code>cwlLoad</code></h3>
<p>The last core function <code>cwlLoad</code> loads the <code>Rcwl</code> tool/pipeline into
the <em>R</em> working environment. The code below loads the tool with a
user-defined name <code>bwa</code> to do the read alignment.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="rcwlpipelines.html#cb113-1"></a>bwa &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="kw">title</span>(t1)[<span class="dv">1</span>])  <span class="co">## &quot;tl_bwa&quot;</span></span>
<span id="cb113-2"><a href="rcwlpipelines.html#cb113-2"></a>bwa &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="kw">mcols</span>(t1)<span class="op">$</span>fpath[<span class="dv">1</span>]) <span class="co">## equivalent to the above. </span></span>
<span id="cb113-3"><a href="rcwlpipelines.html#cb113-3"></a>bwa</span></code></pre></div>
<pre><code>## class: cwlProcess 
##  cwlClass: CommandLineTool 
##  cwlVersion: v1.0 
##  baseCommand: bwa mem 
## requirements:
## - class: DockerRequirement
##   dockerPull: biocontainers/bwa:v0.7.17-3-deb_cv1
## inputs:
##   threads (int): -t 
##   RG (string): -R 
##   Ref (File):  
##   FQ1 (File):  
##   FQ2 (File?):  
## outputs:
## sam:
##   type: File
##   outputBinding:
##     glob: &#39;*.sam&#39;
## stdout: bwaOutput.sam</code></pre>
<p>Now the <em>R</em> tool of <code>bwa</code> is ready to use.</p>
</div>
</div>
<div id="toolpipeline-customization" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> Tool/pipeline customization</h2>
<p>To fit users’ specific needs，the existing tool or pipline can be
easily customized. Here we use the <code>rnaseq_Sf</code> pipeline to demonstrate
how to access and change the arguments of a specific tool inside a
pipeline. This pipeline covers RNA-seq reads quality summary by
<code>fastQC</code>, alignment by <code>STAR</code>, quantification by <code>featureCounts</code> and
quality control by <code>RSeQC</code>.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="rcwlpipelines.html#cb115-1"></a>rnaseq_Sf &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="st">&quot;pl_rnaseq_Sf&quot;</span>)</span></code></pre></div>
<pre><code>## fastqc loaded</code></pre>
<pre><code>## STAR loaded</code></pre>
<pre><code>## sortBam loaded</code></pre>
<pre><code>## samtools_index loaded</code></pre>
<pre><code>## samtools_flagstat loaded</code></pre>
<pre><code>## featureCounts loaded</code></pre>
<pre><code>## gtfToGenePred loaded</code></pre>
<pre><code>## genePredToBed loaded</code></pre>
<pre><code>## read_distribution loaded</code></pre>
<pre><code>## geneBody_coverage loaded</code></pre>
<pre><code>## gCoverage loaded</code></pre>
<pre><code>## STAR loaded</code></pre>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="rcwlpipelines.html#cb128-1"></a><span class="kw">plotCWL</span>(rnaseq_Sf)</span></code></pre></div>
<div id="htmlwidget-0315744de060b156af0e" style="width:672px;height:480px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-0315744de060b156af0e">{"x":{"diagram":"digraph {\n\ngraph [layout = \"neato\",\n       outputorder = \"edgesfirst\",\n       bgcolor = \"white\"]\n\nnode [fontname = \"Helvetica\",\n      fontsize = \"10\",\n      shape = \"circle\",\n      fixedsize = \"true\",\n      width = \"0.5\",\n      style = \"filled\",\n      fillcolor = \"aliceblue\",\n      color = \"gray70\",\n      fontcolor = \"gray50\"]\n\nedge [fontname = \"Helvetica\",\n     fontsize = \"8\",\n     len = \"1.5\",\n     color = \"gray80\",\n     arrowsize = \"0.5\"]\n\n  \"1\" [label = \"in_seqfiles\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"-1,7!\"] \n  \"2\" [label = \"in_prefix\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"0,7!\"] \n  \"3\" [label = \"in_genomeDir\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1.5,7!\"] \n  \"4\" [label = \"in_GTFfile\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"3.5,7!\"] \n  \"5\" [label = \"in_runThreadN\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2.5,7!\"] \n  \"6\" [label = \"out_fastqc\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"-1,5!\"] \n  \"7\" [label = \"out_BAM\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1,3!\"] \n  \"8\" [label = \"out_Log\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"0,5!\"] \n  \"9\" [label = \"out_Count\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"4,5!\"] \n  \"10\" [label = \"out_stat\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,3!\"] \n  \"11\" [label = \"out_count\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"5.5,1!\"] \n  \"12\" [label = \"out_distribution\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"0,1!\"] \n  \"13\" [label = \"out_gCovP\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1,1!\"] \n  \"14\" [label = \"out_gCovT\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2.5,1!\"] \n  \"15\" [label = \"fastqc\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"-1,6!\"] \n  \"16\" [label = \"STAR\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1.5,6!\"] \n  \"17\" [label = \"sortBam\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1.5,5!\"] \n  \"18\" [label = \"samtools_index\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1,4!\"] \n  \"19\" [label = \"samtools_flagstat\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,4!\"] \n  \"20\" [label = \"featureCounts\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"5.5,2!\"] \n  \"21\" [label = \"gtfToGenePred\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"3.5,6!\"] \n  \"22\" [label = \"genePredToBed\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"3,5!\"] \n  \"23\" [label = \"r_distribution\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"0,2!\"] \n  \"24\" [label = \"gCoverage\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,2!\"] \n  \"1\"->\"15\" \n  \"15\"->\"6\" \n  \"2\"->\"16\" \n  \"3\"->\"16\" \n  \"4\"->\"16\" \n  \"1\"->\"16\" \n  \"5\"->\"16\" \n  \"16\"->\"8\" \n  \"16\"->\"9\" \n  \"16\"->\"17\" \n  \"17\"->\"18\" \n  \"18\"->\"7\" \n  \"17\"->\"19\" \n  \"19\"->\"10\" \n  \"4\"->\"20\" \n  \"7\"->\"20\" \n  \"20\"->\"11\" \n  \"4\"->\"21\" \n  \"21\"->\"22\" \n  \"7\"->\"23\" \n  \"22\"->\"23\" \n  \"23\"->\"12\" \n  \"7\"->\"24\" \n  \"22\"->\"24\" \n  \"24\"->\"13\" \n  \"24\"->\"14\" \n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>
<p>There are many default arguments defined for the tool of <code>STAR</code> inside
the pipeline. Users might want to change some of them. For example, we
can change the value for <code>--outFilterMismatchNmax</code> argument from 2 to
5 for longer reads.</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="rcwlpipelines.html#cb129-1"></a><span class="kw">arguments</span>(rnaseq_Sf, <span class="st">&quot;STAR&quot;</span>)[<span class="dv">5</span><span class="op">:</span><span class="dv">6</span>]</span></code></pre></div>
<pre><code>## [[1]]
## [1] &quot;--outFilterMismatchNmax&quot;
## 
## [[2]]
## [1] &quot;2&quot;</code></pre>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="rcwlpipelines.html#cb131-1"></a><span class="kw">arguments</span>(rnaseq_Sf, <span class="st">&quot;STAR&quot;</span>)[[<span class="dv">6</span>]] &lt;-<span class="st"> </span><span class="dv">5</span></span>
<span id="cb131-2"><a href="rcwlpipelines.html#cb131-2"></a><span class="kw">arguments</span>(rnaseq_Sf, <span class="st">&quot;STAR&quot;</span>)[<span class="dv">5</span><span class="op">:</span><span class="dv">6</span>]</span></code></pre></div>
<pre><code>## [[1]]
## [1] &quot;--outFilterMismatchNmax&quot;
## 
## [[2]]
## [1] &quot;5&quot;</code></pre>
<p>We can also change the docker image for a specific tool (e.g., to a
specific version). First, we search for all available docker images
for <code>STAR</code> in biocontainers repository. The Source server could be
<a href="https://quay.io/">quay</a> or <a href="https://hub.docker.com">dockerhub</a>.</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="rcwlpipelines.html#cb133-1"></a><span class="kw">searchContainer</span>(<span class="st">&quot;STAR&quot;</span>, <span class="dt">repo =</span> <span class="st">&quot;biocontainers&quot;</span>, <span class="dt">source =</span> <span class="st">&quot;quay&quot;</span>)</span></code></pre></div>
<pre><code>## DataFrame with 29 rows and 5 columns
##                  tool          repo        name          last_modified
##           &lt;character&gt;   &lt;character&gt; &lt;character&gt;            &lt;character&gt;
## 2.7.8a--0        STAR biocontainers   2.7.8a--0 Sun, 21 Feb 2021 12:..
## 2.7.7a--0        STAR biocontainers   2.7.7a--0 Tue, 29 Dec 2020 13:..
## 2.7.6a--0        STAR biocontainers   2.7.6a--0 Sun, 20 Sep 2020 09:..
## 2.7.5c--0        STAR biocontainers   2.7.5c--0 Mon, 17 Aug 2020 09:..
## 2.7.5b--0        STAR biocontainers   2.7.5b--0 Sat, 01 Aug 2020 17:..
## ...               ...           ...         ...                    ...
## 2.4.0j--0        STAR biocontainers   2.4.0j--0 Tue, 06 Mar 2018 12:..
## 2.5.4a--0        STAR biocontainers   2.5.4a--0 Fri, 26 Jan 2018 21:..
## 2.5.3a--0        STAR biocontainers   2.5.3a--0 Sat, 18 Mar 2017 11:..
## 2.5.2b--0        STAR biocontainers   2.5.2b--0 Tue, 06 Sep 2016 07:..
## 2.5.1b--0        STAR biocontainers   2.5.1b--0 Wed, 11 May 2016 08:..
##                  size
##           &lt;character&gt;
## 2.7.8a--0     7985763
## 2.7.7a--0     7824392
## 2.7.6a--0     7820491
## 2.7.5c--0     7806635
## 2.7.5b--0     7805390
## ...               ...
## 2.4.0j--0     4734325
## 2.5.4a--0     9225952
## 2.5.3a--0     9119736
## 2.5.2b--0     9086803
## 2.5.1b--0    11291827</code></pre>
<p>Then, we can change the <code>STAR</code> version into 2.7.8a (tag name: 2.7.8a–0).</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="rcwlpipelines.html#cb135-1"></a><span class="kw">requirements</span>(rnaseq_Sf, <span class="st">&quot;STAR&quot;</span>)[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## $class
## [1] &quot;DockerRequirement&quot;
## 
## $dockerPull
## [1] &quot;quay.io/biocontainers/star:2.7.3a--0&quot;</code></pre>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="rcwlpipelines.html#cb137-1"></a><span class="kw">requirements</span>(rnaseq_Sf, <span class="st">&quot;STAR&quot;</span>)[[<span class="dv">1</span>]] &lt;-<span class="st"> </span><span class="kw">requireDocker</span>(</span>
<span id="cb137-2"><a href="rcwlpipelines.html#cb137-2"></a>    <span class="dt">docker =</span> <span class="st">&quot;quay.io/biocontainers/star:2.7.8a--0&quot;</span>)</span>
<span id="cb137-3"><a href="rcwlpipelines.html#cb137-3"></a><span class="kw">requirements</span>(rnaseq_Sf, <span class="st">&quot;STAR&quot;</span>)[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## $class
## [1] &quot;DockerRequirement&quot;
## 
## $dockerPull
## [1] &quot;quay.io/biocontainers/star:2.7.8a--0&quot;</code></pre>
</div>
<div id="build-a-pipeline" class="section level2" number="6.4">
<h2><span class="header-section-number">6.4</span> Build a pipeline</h2>
<p>We can build a pipline using the available tools. Here we demonstrate
how to build a simple alignment pipeline with mapping and marking
duplicates.</p>
<p>First, we check whether the required tools (bwa, samtools and picard
markduplicates) are available in <code>RcwlPipelines</code>.</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="rcwlpipelines.html#cb139-1"></a>tls &lt;-<span class="st"> </span><span class="kw">cwlSearch</span>(<span class="st">&quot;bwa|sam2bam|sortBam|samtools_index|markdup&quot;</span>,</span>
<span id="cb139-2"><a href="rcwlpipelines.html#cb139-2"></a>                 <span class="dt">type =</span> <span class="st">&quot;tool&quot;</span>)</span>
<span id="cb139-3"><a href="rcwlpipelines.html#cb139-3"></a>tls</span></code></pre></div>
<pre><code>## cwlHub with 6 records
## cache path:  ~/Library/Caches/Rcwl 
## # last modified date:  2021-02-22 
## # cwlSearch() to query scripts
## # cwlLoad(&#39;title&#39;) to load the script
## # additional mcols(): rid, rpath, Type, Container, mtime, ...
## 
##             title             Command              
##   BFC2814 | tl_bwa_index      bwa index            
##   BFC2815 | tl_bwa            bwa mem              
##   BFC2851 | tl_markdup        picard MarkDuplicates
##   BFC2877 | tl_sam2bam        samtools view        
##   BFC2881 | tl_samtools_index samtools index       
##   BFC2887 | tl_sortBam        samtools sort</code></pre>
<p>Then we load all the required tools.</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="rcwlpipelines.html#cb141-1"></a>bwa &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="st">&quot;tl_bwa&quot;</span>)</span>
<span id="cb141-2"><a href="rcwlpipelines.html#cb141-2"></a>bwa_index &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="st">&quot;tl_bwa_index&quot;</span>)</span>
<span id="cb141-3"><a href="rcwlpipelines.html#cb141-3"></a>markdup &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="st">&quot;tl_markdup&quot;</span>)</span>
<span id="cb141-4"><a href="rcwlpipelines.html#cb141-4"></a>sam2bam &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="st">&quot;tl_sam2bam&quot;</span>)</span>
<span id="cb141-5"><a href="rcwlpipelines.html#cb141-5"></a>samtools_index &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="st">&quot;tl_samtools_index&quot;</span>)</span>
<span id="cb141-6"><a href="rcwlpipelines.html#cb141-6"></a>sortBam &lt;-<span class="st"> </span><span class="kw">cwlLoad</span>(<span class="st">&quot;tl_sortBam&quot;</span>)</span></code></pre></div>
<p>Next, we will need to define the input parameters for the pipeline
(instead of for each tool).</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="rcwlpipelines.html#cb142-1"></a>p1 &lt;-<span class="st"> </span><span class="kw">InputParam</span>(<span class="dt">id =</span> <span class="st">&quot;threads&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;int&quot;</span>)</span>
<span id="cb142-2"><a href="rcwlpipelines.html#cb142-2"></a>p2 &lt;-<span class="st"> </span><span class="kw">InputParam</span>(<span class="dt">id =</span> <span class="st">&quot;RG&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;string&quot;</span>)</span>
<span id="cb142-3"><a href="rcwlpipelines.html#cb142-3"></a>p3 &lt;-<span class="st"> </span><span class="kw">InputParam</span>(<span class="dt">id =</span> <span class="st">&quot;Ref&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;string&quot;</span>)</span>
<span id="cb142-4"><a href="rcwlpipelines.html#cb142-4"></a>p4 &lt;-<span class="st"> </span><span class="kw">InputParam</span>(<span class="dt">id =</span> <span class="st">&quot;FQ1&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;File&quot;</span>)</span>
<span id="cb142-5"><a href="rcwlpipelines.html#cb142-5"></a>p5 &lt;-<span class="st"> </span><span class="kw">InputParam</span>(<span class="dt">id =</span> <span class="st">&quot;FQ2&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;File?&quot;</span>)</span></code></pre></div>
<p>Then we define the pipeline steps, to connect the inputs and outputs
of each tool to form a pipeline.</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="rcwlpipelines.html#cb143-1"></a><span class="co">## bwa</span></span>
<span id="cb143-2"><a href="rcwlpipelines.html#cb143-2"></a>s1 &lt;-<span class="st"> </span><span class="kw">cwlStep</span>(<span class="dt">id =</span> <span class="st">&quot;bwa&quot;</span>, <span class="dt">run =</span> bwa,</span>
<span id="cb143-3"><a href="rcwlpipelines.html#cb143-3"></a>              <span class="dt">In =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="st">&quot;threads&quot;</span>,</span>
<span id="cb143-4"><a href="rcwlpipelines.html#cb143-4"></a>                        <span class="dt">RG =</span> <span class="st">&quot;RG&quot;</span>,</span>
<span id="cb143-5"><a href="rcwlpipelines.html#cb143-5"></a>                        <span class="dt">Ref =</span> <span class="st">&quot;Ref&quot;</span>,</span>
<span id="cb143-6"><a href="rcwlpipelines.html#cb143-6"></a>                        <span class="dt">FQ1 =</span> <span class="st">&quot;FQ1&quot;</span>,</span>
<span id="cb143-7"><a href="rcwlpipelines.html#cb143-7"></a>                        <span class="dt">FQ2 =</span> <span class="st">&quot;FQ2&quot;</span>))</span>
<span id="cb143-8"><a href="rcwlpipelines.html#cb143-8"></a><span class="co">## sam to bam</span></span>
<span id="cb143-9"><a href="rcwlpipelines.html#cb143-9"></a>s2 &lt;-<span class="st"> </span><span class="kw">cwlStep</span>(<span class="dt">id =</span> <span class="st">&quot;sam2bam&quot;</span>, <span class="dt">run =</span> sam2bam,</span>
<span id="cb143-10"><a href="rcwlpipelines.html#cb143-10"></a>              <span class="dt">In =</span> <span class="kw">list</span>(<span class="dt">sam =</span> <span class="st">&quot;bwa/sam&quot;</span>))</span>
<span id="cb143-11"><a href="rcwlpipelines.html#cb143-11"></a><span class="co">## sort bam</span></span>
<span id="cb143-12"><a href="rcwlpipelines.html#cb143-12"></a>s3 &lt;-<span class="st"> </span><span class="kw">cwlStep</span>(<span class="dt">id =</span> <span class="st">&quot;sortBam&quot;</span>, <span class="dt">run =</span> sortBam,</span>
<span id="cb143-13"><a href="rcwlpipelines.html#cb143-13"></a>              <span class="dt">In =</span> <span class="kw">list</span>(<span class="dt">bam =</span> <span class="st">&quot;sam2bam/bam&quot;</span>))</span>
<span id="cb143-14"><a href="rcwlpipelines.html#cb143-14"></a><span class="co">## mark duplicates</span></span>
<span id="cb143-15"><a href="rcwlpipelines.html#cb143-15"></a>s4 &lt;-<span class="st"> </span><span class="kw">cwlStep</span>(<span class="dt">id =</span> <span class="st">&quot;markdup&quot;</span>, <span class="dt">run =</span> markdup,</span>
<span id="cb143-16"><a href="rcwlpipelines.html#cb143-16"></a>              <span class="dt">In =</span> <span class="kw">list</span>(<span class="dt">ibam =</span> <span class="st">&quot;sortBam/sbam&quot;</span>,</span>
<span id="cb143-17"><a href="rcwlpipelines.html#cb143-17"></a>                        <span class="dt">obam =</span> <span class="kw">list</span>(</span>
<span id="cb143-18"><a href="rcwlpipelines.html#cb143-18"></a>                            <span class="dt">valueFrom=</span><span class="st">&quot;$(inputs.ibam.nameroot).mdup.bam&quot;</span>),</span>
<span id="cb143-19"><a href="rcwlpipelines.html#cb143-19"></a>                        <span class="dt">matrix =</span> <span class="kw">list</span>(</span>
<span id="cb143-20"><a href="rcwlpipelines.html#cb143-20"></a>                            <span class="dt">valueFrom=</span><span class="st">&quot;$(inputs.ibam.nameroot).markdup.txt&quot;</span>)))</span>
<span id="cb143-21"><a href="rcwlpipelines.html#cb143-21"></a><span class="co">## index bam</span></span>
<span id="cb143-22"><a href="rcwlpipelines.html#cb143-22"></a>s5 &lt;-<span class="st"> </span><span class="kw">cwlStep</span>(<span class="dt">id =</span> <span class="st">&quot;idxBam&quot;</span>, <span class="dt">run =</span> samtools_index,</span>
<span id="cb143-23"><a href="rcwlpipelines.html#cb143-23"></a>              <span class="dt">In =</span> <span class="kw">list</span>(<span class="dt">bam =</span> <span class="st">&quot;markdup/mBam&quot;</span>))</span></code></pre></div>
<p>Last, we will define the pipeline outputs and connect all the above
defined steps into a new pipeline.</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="rcwlpipelines.html#cb144-1"></a>req1 &lt;-<span class="st"> </span><span class="kw">requireStepInputExpression</span>()</span>
<span id="cb144-2"><a href="rcwlpipelines.html#cb144-2"></a>req2 &lt;-<span class="st"> </span><span class="kw">requireJS</span>()</span>
<span id="cb144-3"><a href="rcwlpipelines.html#cb144-3"></a><span class="co">## outputs</span></span>
<span id="cb144-4"><a href="rcwlpipelines.html#cb144-4"></a>o1 &lt;-<span class="st"> </span><span class="kw">OutputParam</span>(<span class="dt">id =</span> <span class="st">&quot;Bam&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;File&quot;</span>, <span class="dt">outputSource =</span> <span class="st">&quot;markdup/mBam&quot;</span>)</span>
<span id="cb144-5"><a href="rcwlpipelines.html#cb144-5"></a>o2 &lt;-<span class="st"> </span><span class="kw">OutputParam</span>(<span class="dt">id =</span> <span class="st">&quot;Idx&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;File&quot;</span>, <span class="dt">outputSource =</span> <span class="st">&quot;idxBam/idx&quot;</span>)</span>
<span id="cb144-6"><a href="rcwlpipelines.html#cb144-6"></a><span class="co">## cwlWorkflow</span></span>
<span id="cb144-7"><a href="rcwlpipelines.html#cb144-7"></a>Align &lt;-<span class="st"> </span><span class="kw">cwlWorkflow</span>(<span class="dt">requirements =</span> <span class="kw">list</span>(req1, req2),</span>
<span id="cb144-8"><a href="rcwlpipelines.html#cb144-8"></a>                     <span class="dt">inputs =</span> <span class="kw">InputParamList</span>(p1, p2, p3, p4, p5),</span>
<span id="cb144-9"><a href="rcwlpipelines.html#cb144-9"></a>                     <span class="dt">outputs =</span> <span class="kw">OutputParamList</span>(o1, o2))</span>
<span id="cb144-10"><a href="rcwlpipelines.html#cb144-10"></a><span class="co">## build pipeline</span></span>
<span id="cb144-11"><a href="rcwlpipelines.html#cb144-11"></a>Align &lt;-<span class="st"> </span>Align <span class="op">+</span><span class="st"> </span>s1 <span class="op">+</span><span class="st"> </span>s2 <span class="op">+</span><span class="st"> </span>s3 <span class="op">+</span><span class="st"> </span>s4 <span class="op">+</span><span class="st"> </span>s5</span></code></pre></div>
<p>Now the pipeline is successfully built and ready for use. We can
visualize the pipeline with <code>plotCWL</code> from the <code>Rcwl</code> package.</p>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="rcwlpipelines.html#cb145-1"></a><span class="kw">plotCWL</span>(Align)</span></code></pre></div>
<div id="htmlwidget-3bffe57c7a9c3442edac" style="width:672px;height:480px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-3bffe57c7a9c3442edac">{"x":{"diagram":"digraph {\n\ngraph [layout = \"neato\",\n       outputorder = \"edgesfirst\",\n       bgcolor = \"white\"]\n\nnode [fontname = \"Helvetica\",\n      fontsize = \"10\",\n      shape = \"circle\",\n      fixedsize = \"true\",\n      width = \"0.5\",\n      style = \"filled\",\n      fillcolor = \"aliceblue\",\n      color = \"gray70\",\n      fontcolor = \"gray50\"]\n\nedge [fontname = \"Helvetica\",\n     fontsize = \"8\",\n     len = \"1.5\",\n     color = \"gray80\",\n     arrowsize = \"0.5\"]\n\n  \"1\" [label = \"threads\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"0,8!\"] \n  \"2\" [label = \"RG\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1,8!\"] \n  \"3\" [label = \"Ref\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,8!\"] \n  \"4\" [label = \"FQ1\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"3,8!\"] \n  \"5\" [label = \"FQ2\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"4,8!\"] \n  \"6\" [label = \"Bam\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,3!\"] \n  \"7\" [label = \"Idx\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,1!\"] \n  \"8\" [label = \"bwa\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,7!\"] \n  \"9\" [label = \"sam2bam\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,6!\"] \n  \"10\" [label = \"sortBam\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,5!\"] \n  \"11\" [label = \"markdup\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,4!\"] \n  \"12\" [label = \"idxBam\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,2!\"] \n  \"1\"->\"8\" \n  \"2\"->\"8\" \n  \"3\"->\"8\" \n  \"4\"->\"8\" \n  \"5\"->\"8\" \n  \"8\"->\"9\" \n  \"9\"->\"10\" \n  \"10\"->\"11\" \n  \"11\"->\"6\" \n  \"6\"->\"12\" \n  \"12\"->\"7\" \n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="toolpipeline-execution.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="dnaseq-alignment.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/06-RcwlPipelines.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Rcwl-Pipelines.pdf", "Rcwl-Pipelines.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
